# The Armenian Name Inventor - a neural net learns to speak surname

**Instructor-led practical, chapter 11 (neural networks).** Builds on [41] (the network),
[42]-[43] (training it, early stopping), [44] (the Adam optimizer) and [46] (the same
machinery written from scratch in numpy).

Every family tree of ChatGPT leads back to the classic neural language model of
[Bengio et al., 2003](https://www.jmlr.org/papers/volume3/bengio03a/bengio03a.pdf) - and that
model was nothing but an **embedding layer plus an MLP**. Exactly the parts we built in [41]
and trained in [42]-[43]. No attention, no recurrence, no magic.

Today we train that model on **689 Armenian surnames** scraped from the Armenian Wikipedia,
and then ask it to **invent surnames that do not exist**. Two tools are new:

- **PyTorch** - the deep-learning library the whole field runs on. First contact; every new
  piece gets explained the moment we use it.
- **Weights & Biases (wandb)** - an experiment tracker: a lab notebook that writes itself.
  Today involves fourteen trainings - ten of them in one comparison - and nobody can hold
  fourteen loss curves in their head.

**The plan**

1. The data, 2. the alphabet, 3. context windows - and why the window is 3 letters
4. The model: an embedding (opened up: it is just a weight matrix) plus an MLP - and the same
   model without an embedding layer
5. Sampling: asking the net to write, one letter at a time
6. How good is good? Loss, perplexity, and two baselines to beat
7. Training, logged to wandb
8. The payoff
9. Which knob matters? A small comparison
10. Saving the model and loading it back
11. - 14. Bonuses: temperature, memorization, a map of the letters, steering

**Runtime:** a full rerun takes ~6-9 min on the instructor's laptop CPU, depending on how
busy it is. Slow cells say so on their first line; the big ones are the live training in section 7 (~2 min, most of it
deliberate pauses and wandb start-up), the UMAP map in bonus 13 (~1.5-3 min, almost all of it
library start-up), the comparison in section 9 (~1-2 min) and the memorization run in bonus 12
(~1 min).

Seed is fixed to `509`. Everything runs on a laptop CPU.

In [1]:
# ~15-25 s (importing torch and wandb)
import copy
import os
import time
from pathlib import Path

import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import wandb

# install into the course venv if missing:  uv pip install torch wandb plotly umap-learn
SEED = 509
torch.manual_seed(SEED)

# This model is tiny (~8k weights). Splitting such small matrix products across CPU threads
# costs more than it saves: measured on the instructor laptop, one thread trains an epoch in
# 66 ms vs 130 ms with torch's default. For big models it is the opposite - keep the default.
torch.set_num_threads(1)

# plot colors (the Armenian flag, plus gray for reference lines)
RED, BLUE, ORANGE, GRAY = "#D90012", "#0033A0", "#F2A800", "#888888"

print("torch", torch.__version__, "| wandb", wandb.__version__)

torch 2.9.0+cpu | wandb 0.28.2


In [2]:
# Credentials hygiene: an API key never goes into a notebook (notebooks get committed,
# shared, screenshotted). Ours lives in the repo-root .env (gitignored) as
# WANDB_API_KEY=... - loaded here if present; otherwise wandb falls back to `wandb login`.
env_file = Path("../../.env")
if env_file.exists():
    for line in env_file.read_text(encoding="utf-8").splitlines():
        key, _, value = line.partition("=")
        if key == "WANDB_API_KEY" and value.strip():
            os.environ["WANDB_API_KEY"] = value.strip()

print("WANDB_API_KEY is set:", "WANDB_API_KEY" in os.environ)

WANDB_API_KEY is set: True


## 1. The data: every Armenian surname Wikipedia knows

The Armenian Wikipedia keeps a category
[«Հայկական ազգանուններ»](https://hy.wikipedia.org/wiki/Կատեգորիա:Հայկական_ազգանուններ) with
one page per surname. The prep script `py_src/fetch_surnames.py` pulled every page title,
kept only titles made purely of Armenian letters (dropping disambiguation pages like
«Ազարյան (ազգանուն)» and hyphenated forms), lowercased them so capitals do not double our
alphabet, and pinned the result to `data/surnames_hy.txt` - so this notebook needs no
internet.

In [3]:
names = open("data/surnames_hy.txt", encoding="utf-8").read().split()

lengths = [len(n) for n in names]
n_yan = sum(n.endswith("յան") for n in names)
print(f"surnames: {len(names)}")
print(f"length min/mean/max: {min(lengths)}/{np.mean(lengths):.1f}/{max(lengths)}")
print(f"ending in -յան: {n_yan} ({n_yan / len(names):.0%})")
print("a few:", ", ".join(names[100:108]))

surnames: 689
length min/mean/max: 5/8.5/14
ending in -յան: 660 (96%)
a few: անանյան, անասյան, անդրանիկյան, անդրեասյան, անդրիասյան, անթառամյան, անջելյան, անտիկյան


## 2. The alphabet, discovered from the data

**Predict first:** how many *distinct characters* do 689 surnames use? (The Armenian
alphabet has 39 letters - do surnames use all of them?)

We do not hardcode the alphabet; we read it off the data and number the letters with two
lookup tables. Their names are the standard ones from Karpathy's *makemore*:

- **`stoi`** = **s**tring **to** **i**nteger: letter -> code, e.g. `stoi["պ"] = 25`
- **`itos`** = **i**nteger **to** **s**tring: code -> letter, e.g. `itos[25] = "պ"`

We also add two special symbols, one for each end of a name:

- **`^` - start.** It pads the context before the first letter: `^^^` means "nothing written
  yet". The model **reads** it but never **writes** it.
- **`.` - end.** It is the answer "stop": this is how the model learns that «-յան» is a way to
  finish. The model **writes** it but never **reads** it - once `.` appears, the name is over.

(LLMs have the same pair, usually called BOS and EOS - beginning and end of sequence.)

So the model has two vocabularies:

- what it can **write** - the softmax at the output: the 38 letters and `.`, so
  **$V_{out} = 39$**;
- what it can **read** - the rows of the embedding table at the input: every symbol with a
  code, so **$V_{in} = 40$**. That includes a row for `.`, which is never read, so it never
  trains: 8 idle weights, the price of numbering both directions with one `stoi`.

`.` gets code 0 and `^` the last code, 39, so the 39 writable symbols are exactly the codes
0-38 - the 39 outputs of the network.

In [4]:
chars = sorted(set("".join(names)))
stoi = {c: i + 1 for i, c in enumerate(chars)}   # "string to integer": letter -> code 1..38
stoi["."] = 0                                    # end: written, never read
stoi["^"] = len(chars) + 1                       # start / padding: read, never written (39)
itos = {i: c for c, i in stoi.items()}           # "integer to string": code -> symbol
PAD = stoi["^"]
V_in = len(stoi)                                 # symbols the model can read: embedding rows
V_out = len(stoi) - 1                            # symbols it can write: every code except ^

print("alphabet found in the data:", "".join(chars))
print(f"letters found: {len(chars)}")
print(f"V_in  = {V_in}  (38 letters + '.' + '^': rows of the embedding table)")
print(f"V_out = {V_out}  (38 letters + '.': outputs of the softmax)")

armenian = {chr(c) for c in range(ord("ա"), ord("և") + 1)}   # all lowercase Armenian letters
print("Armenian letters that never occur in a surname:", sorted(armenian - set(chars)))

alphabet found in the data: աբգդեզէթժիլխծկհձղճմյնշոչպջռսվտրցւփքօֆև
letters found: 38
V_in  = 40  (38 letters + '.' + '^': rows of the embedding table)
V_out = 39  (38 letters + '.': outputs of the softmax)
Armenian letters that never occur in a surname: ['ը']


**ը never occurs** in the 689 surnames, so it gets no code: the model can never write it, and
it cannot read it either. Any character outside the alphabet is *out of vocabulary*. The
helper below turns text into codes and **refuses loudly** anything it has no code for.
Silently skipping the character would quietly change what the model is asked - you would get
an answer to a question you did not pose. (Section 14 shows the error.) Most modern LLMs
avoid the problem at the root: their vocabulary contains all 256 possible byte values, so any
text in any script can always be encoded.

In [5]:
def encode(text):
    """Letters -> integer codes. Raises on any character the model has no code for."""
    text = text.lower()                          # the names were lowercased too
    unknown = sorted(set(text) - set(stoi))
    if unknown:
        raise ValueError(f"{unknown} never appear in the 689 surnames, so the model has no "
                         f"code (and no embedding row) for them. Its alphabet: {''.join(chars)}")
    return [stoi[c] for c in text]


print("encode('Պող') ->", encode("Պող"))

encode('Պող') -> [25, 23, 17]


We hold out 20% of the **names** for validation. By names, not by training windows - windows
cut from the same name are near-duplicates, and splitting them across train/val would leak
(the same trap as ch2's rent model and the [06] evaluation lecture).

In [6]:
rng = np.random.default_rng(SEED)
perm = rng.permutation(len(names))
n_tr = int(0.8 * len(names))
tr_names = [names[i] for i in perm[:n_tr]]
va_names = [names[i] for i in perm[n_tr:]]
tr_set, va_set = set(tr_names), set(va_names)   # for "is this a real name?" checks later
print(f"train names: {len(tr_names)}, val names: {len(va_names)}")

train names: 551, val names: 138


## 3. From characters to model food

**The trap first.** The obvious encoding - Ա=1, Բ=2, ... - lies to the model: it says Բ is
"twice" Ա and that Դ is "between" Գ and Ե. None of that is real. Characters are
**categories**, and we learned the honest encoding for categories back in the
feature-engineering lecture: **one-hot** (all zeros, a single 1). We will store integer
*codes*, but only as row numbers into one-hots - never as quantities (section 4 shows
exactly how).

**The task shape.** The model reads a **context window** of the previous $K$ characters and
predicts the next one. Sliding that window over «պողոսյան» with $K = 3$:

```
^ ^ ^  ->  պ          ո ս յ  ->  ա
^ ^ պ  ->  ո          ս յ ա  ->  ն
^ պ ո  ->  ղ          յ ա ն  ->  .   <- "now stop"
պ ո ղ  ->  ո  ...
```

One name of length $L$ yields $L{+}1$ training examples. The whole dataset is just these
(context, next-char) pairs.

In [7]:
def build_dataset(name_list, K):
    """Slide a K-wide window over every name; return (contexts X, next-chars Y)."""
    X, Y = [], []
    for n in name_list:
        ctx = [PAD] * K                    # start: K start tokens (^) as padding
        for ch in n + ".":                 # append '.' so the net learns to stop
            X.append(ctx)
            Y.append(stoi[ch])
            ctx = ctx[1:] + [stoi[ch]]     # slide the window one letter to the right
    return torch.tensor(X), torch.tensor(Y)


def decode(row):
    """Integer codes back to letters, e.g. tensor([25, 23, 17]) -> 'պող'."""
    return "".join(itos[int(i)] for i in row)

**Two minutes on tensors,** exactly where we first need them: a `torch.tensor` is a numpy
array that (a) remembers how it was computed, so gradients can flow backward through it
(that is autograd - the machinery running [42]'s backprop for us), and (b) can live on a GPU.
The two things you always check are `.shape` and `.dtype`.

In [8]:
K = 3
Xtr, Ytr = build_dataset(tr_names, K)
Xva, Yva = build_dataset(va_names, K)
print(f"train windows: {tuple(Xtr.shape)}, targets: {tuple(Ytr.shape)}, dtype: {Xtr.dtype}")
print(f"val windows:   {tuple(Xva.shape)}")

first = tr_names[0]
print(f"\nthe {len(first) + 1} windows cut from the first training name, «{first}»:")
for i in range(len(first) + 1):
    print(f"  '{decode(Xtr[i])}' -> '{itos[int(Ytr[i])]}'")

train windows: (5208, 3), targets: (5208,), dtype: torch.int64
val windows:   (1322, 3)

the 9 windows cut from the first training name, «մեխակյան»:
  '^^^' -> 'մ'
  '^^մ' -> 'ե'
  '^մե' -> 'խ'
  'մեխ' -> 'ա'
  'եխա' -> 'կ'
  'խակ' -> 'յ'
  'ակյ' -> 'ա'
  'կյա' -> 'ն'
  'յան' -> '.'


The promise from section 2 - `^` is only ever read, `.` is only ever written - checked on
all 5,208 training windows:

In [9]:
print(f"'^' as an answer:      {int((Ytr == stoi['^']).sum())} times  (it can never be written)")
print(f"'.' inside a context:  {int((Xtr == stoi['.']).sum())} times  (nothing is read after the end)")
print(f"'.' as an answer:      {int((Ytr == stoi['.']).sum())} times  "
      f"= once per training name ({len(tr_names)})")

'^' as an answer:      0 times  (it can never be written)


'.' inside a context:  0 times  (nothing is read after the end)
'.' as an answer:      551 times  = once per training name (551)


### Why K = 3?

$K$ is how many letters the model gets to look back. Two simple facts pin it down.

**Fact 1: to know that a name is over, the model has to see «յան».** 96% of the names end in
«-յան», so the most important decision the model makes is *when to stop*. Put yourself in
its place: you have just written «...ն» and may look back only $K$ letters. Is the name over?

The cell below finds every place the last 1, 2 or 3 letters occur in the training names,
counts how often the name really ends right there, and shows a few names where it does not
(the matched letters in brackets):

In [10]:
def where_it_occurs(context):
    """Every occurrence of `context` in the training names, split into 'name ends here' and
    'name goes on', with the occurrence marked in brackets."""
    ends, goes_on = [], []
    for name in tr_names:
        for pos in range(len(name) - len(context) + 1):
            if name[pos:pos + len(context)] == context:
                marked = name[:pos] + "[" + context + "]" + name[pos + len(context):]
                (ends if pos + len(context) == len(name) else goes_on).append(marked)
    return ends, goes_on


p_stop = {}
for context in ["ն", "ան", "յան"]:
    ends, goes_on = where_it_occurs(context)
    p_stop[context] = len(ends) / (len(ends) + len(goes_on))
    print(f"last {len(context)} letter(s) = '{context}': seen {len(ends) + len(goes_on)} times, "
          f"the name ends there {len(ends)} times ({p_stop[context]:.0%})")
    print(f"   it goes on in e.g. {', '.join(goes_on[:4])}\n")

fig = go.Figure(go.Bar(
    x=[f"K={len(c)}: sees '{c}'" for c in p_stop], y=list(p_stop.values()), marker_color=BLUE,
    text=[f"{p:.0%}" for p in p_stop.values()], textposition="outside", cliponaxis=False))
fig.update_layout(title="How sure can the model be that the name is over?",
                  yaxis=dict(title="P(next is '.' = stop)", tickformat=".0%", range=[0, 1.12]),
                  width=620, height=360, template="plotly_white")
fig.show()

last 1 letter(s) = 'ն': seen 716 times, the name ends there 527 times (74%)
   it goes on in e.g. օզա[ն]յան, ռշտու[ն]ի, սողոմո[ն]յան, գուրգե[ն]յան

last 2 letter(s) = 'ան': seen 638 times, the name ends there 527 times (83%)
   it goes on in e.g. օզ[ան]յան, ադ[ան]ալյան, փարսադ[ան]յան, ս[ան]ամյան

last 3 letter(s) = 'յան': seen 534 times, the name ends there 527 times (99%)
   it goes on in e.g. բարաթ[յան]ց, բադամ[յան]ց, ավագ[յան]ց, ա[յան]յան



With one letter the model cannot tell «...ն» at the end from ն in the middle of a stem. With
two, «ան» still shows up mid-name. With three, «յան» is almost always the end - the few
exceptions are names that continue past it (like the «-յանց» ones in the list). **So K should
be at least 3.**

**Fact 2: every extra letter makes the data thinner.** Each letter of context multiplies the
number of possible contexts by 39, but we still have only 551 training names. So the longer
the window, the more often the model meets a context it has *never seen in training* - and
for those it has nothing specific to go on. First one validation name, window by window
(`NEW` = this exact context never occurs in the training names):

In [11]:
def seen_contexts(k):
    """Every K-letter context that occurs in the training windows."""
    Xk, _ = build_dataset(tr_names, k)
    return {tuple(row) for row in Xk.tolist()}


example = va_names[0]
for k in [3, 5]:
    Xk_ex, Yk_ex = build_dataset([example], k)
    seen = seen_contexts(k)
    steps = [f"'{decode(x)}'->{itos[int(y)]}" + ("" if tuple(x.tolist()) in seen else " NEW")
             for x, y in zip(Xk_ex, Yk_ex)]
    n_new = sum(s.endswith("NEW") for s in steps)
    print(f"«{example}», K={k}: {n_new} of {len(steps)} windows are new")
    print("   " + "  ".join(steps) + "\n")

«կազումյան», K=3: 1 of 10 windows are new
   '^^^'->կ  '^^կ'->ա  '^կա'->զ  'կազ'->ո NEW  'ազո'->ւ  'զու'->մ  'ում'->յ  'ւմյ'->ա  'մյա'->ն  'յան'->.

«կազումյան», K=5: 5 of 10 windows are new
   '^^^^^'->կ  '^^^^կ'->ա  '^^^կա'->զ  '^^կազ'->ո NEW  '^կազո'->ւ NEW  'կազու'->մ NEW  'ազում'->յ NEW  'զումյ'->ա NEW  'ումյա'->ն  'ւմյան'->.



And the same count over all 1,322 validation windows, for every K:

In [12]:
n_slot = len(chars) + 1                    # each context slot holds ^ or one of the 38 letters
Ks, unseen_share = [1, 2, 3, 4, 5, 6, 8], []
print(f"{'K':>2} {'possible contexts':>22} {'seen in training':>17} {'val windows, unseen context':>28}")
for k in Ks:
    Xk_va, _ = build_dataset(va_names, k)
    seen = seen_contexts(k)
    unseen_share.append(np.mean([tuple(row) not in seen for row in Xk_va.tolist()]))
    print(f"{k:>2} {n_slot ** k:>22,} {len(seen):>17,} {unseen_share[-1]:>28.1%}")

fig = go.Figure(go.Bar(x=[f"K={k}" for k in Ks], y=unseen_share, marker_color=RED,
                       text=[f"{u:.0%}" for u in unseen_share], textposition="outside",
                       cliponaxis=False))
fig.update_layout(title="Validation windows whose context never appears in training",
                  yaxis=dict(title="share of validation windows", tickformat=".0%", range=[0, 0.8]),
                  width=620, height=360, template="plotly_white")
fig.show()

 K      possible contexts  seen in training  val windows, unseen context
 1                     39                39                         0.0%
 2                  1,521               422                         3.6%


 3                 59,319             1,313                        16.5%
 4              2,313,441             2,147                        32.5%
 5             90,224,199             2,730                        44.7%
 6          3,518,743,761             3,126                        53.9%
 8      5,352,009,260,481             3,365                        61.6%


At K = 3 one validation window in six is already new; at K = 5 almost half are; at K = 8 most
are. A model that just counts contexts would have nothing to say about those. The MLP copes
better: it never looks up a whole context as one unit - its first layer adds up one
contribution per letter-in-position - so a new combination of familiar letters still gets a
sensible answer. But the extra letters come in combinations it has barely met, so there is
little left for it to learn from them. **So K should be as small as possible.**

Together: **K = 3 is the smallest window that sees the whole suffix.** That is the whole
argument - and the same choice Karpathy's *makemore* makes for 32,000 English names. It is a
design argument, not a proof: section 9 trains K = 1, 2, 3, 4, 5, 6 and 8 and measures what
each extra letter actually buys.

## 4. The model: an embedding and an MLP

### 4.1 The embedding layer is just a weight matrix

Start from what we know: a letter as a one-hot vector. Feed a one-hot into a linear layer,
i.e. multiply it by a weight matrix $W$ (one row per letter). The zeros wipe out every row of
$W$ except one:

$$\text{one-hot}(i) \cdot W = 0 \cdot W_0 + \dots + 1 \cdot W_i + \dots + 0 \cdot W_{V-1} = W_i
\quad \text{(row } i \text{ of } W\text{)}$$

So "multiply the one-hot by $W$" and "look up row $i$ of $W$" are **the same operation**; the
lookup just skips multiplying by all those zeros. That lookup table is what `nn.Embedding`
is. First with numbers small enough to check by eye - a 4-letter alphabet, 2 numbers per
letter:

In [13]:
toy_letters = ["ա", "բ", "գ", "դ"]
W = torch.tensor([[ 0.1,  0.2],     # row 0: the vector for ա
                  [ 1.0, -1.0],     # row 1: բ
                  [ 3.0,  4.0],     # row 2: գ
                  [-5.0,  0.5]])    # row 3: դ

onehot_g = torch.tensor([0.0, 0.0, 1.0, 0.0])          # one-hot for գ (position 2)
print("one-hot for գ:            ", onehot_g.tolist())
print("one-hot @ W  (multiply):  ", (onehot_g @ W).tolist())
print("W[2]         (look up):   ", W[2].tolist())

one-hot for գ:             [0.0, 0.0, 1.0, 0.0]
one-hot @ W  (multiply):   [3.0, 4.0]
W[2]         (look up):    [3.0, 4.0]


Now the real one. `nn.Embedding(V_in, d)` holds a single matrix of shape $(V_{in}, d)$: 40
rows, one per readable symbol, $d$ numbers each. It is a trainable weight like any other -
backprop updates it. Step by step for «պող», first the one-hots:

In [14]:
emb = nn.Embedding(V_in, 8)                            # the lookup table: 40 rows x 8 columns
print("emb.weight shape:", tuple(emb.weight.shape), "| trainable:", emb.weight.requires_grad)

codes = torch.tensor(encode("պող"))                   # 3 letters -> 3 row numbers
print("codes for 'պող':", codes.tolist())

onehots = F.one_hot(codes, num_classes=V_in).float()   # 3 one-hots             -> (3, 40)
print(f"\nthe 3 one-hots, {V_in} numbers each:")
for c, row in zip("պող", onehots.int().tolist()):
    print(f"  {c}: {''.join(map(str, row))}   <- the single 1 sits at position {row.index(1)}")

emb.weight shape: (40, 8) | trainable: True
codes for 'պող': [25, 23, 17]

the 3 one-hots, 40 numbers each:
  պ: 0000000000000000000000000100000000000000   <- the single 1 sits at position 25
  ո: 0000000000000000000000010000000000000000   <- the single 1 sits at position 23
  ղ: 0000000000000000010000000000000000000000   <- the single 1 sits at position 17


Then both routes to the same 3 x 8 result - multiply the one-hots by the whole table, or just
look up rows 25, 23 and 17:

In [15]:
by_matmul = onehots @ emb.weight                       # (3, 40) @ (40, 8)      -> (3, 8)
by_lookup = emb(codes)                                 # rows 25, 23, 17        -> (3, 8)
print("one-hots @ table:\n", by_matmul.detach().round(decimals=2))
print("\nlookup of rows 25, 23, 17:\n", by_lookup.detach().round(decimals=2))
print("\nidentical:", torch.allclose(by_lookup, by_matmul))

one-hots @ table:
 tensor([[-2.2400, -0.0400, -0.7000,  0.0500,  0.2800, -0.5700, -0.2800, -0.6300],
        [-0.2000,  0.1900, -0.5800, -1.3900, -0.2300, -1.1000,  0.2200, -1.2400],
        [ 0.4600,  0.9500,  0.7100,  1.6000, -0.6000, -0.9300, -0.7500, -0.8200]])

lookup of rows 25, 23, 17:
 tensor([[-2.2400, -0.0400, -0.7000,  0.0500,  0.2800, -0.5700, -0.2800, -0.6300],
        [-0.2000,  0.1900, -0.5800, -1.3900, -0.2300, -1.1000,  0.2200, -1.2400],
        [ 0.4600,  0.9500,  0.7100,  1.6000, -0.6000, -0.9300, -0.7500, -0.8200]])

identical: True


**Which rows does training change?** To get gradients, backprop needs one number to
differentiate - a loss. In real training that is the cross-entropy of section 6; here we
invent the simplest stand-in: **add up the 24 numbers we just looked up**. Each looked-up
number appears in that sum once, so its gradient is exactly 1. Every number the lookup never
touched does not appear in the sum at all, so its gradient is 0. A real loss gives other
values than 1 - but the zeros land in exactly the same places.

In [16]:
fake_loss = by_lookup.sum()                    # 3 rows x 8 numbers, added up -> one number
print(f"stand-in loss = sum of the 24 looked-up numbers = {fake_loss.item():.3f}")
fake_loss.backward()                           # fills emb.weight.grad, same shape as the table (40, 8)

grad = emb.weight.grad
print("\ngradient of the rows that were looked up:")
for c in "պող":
    print(f"  row {stoi[c]:>2} ({c}): {grad[stoi[c]].tolist()}")
print("gradient of two rows that were not:")
for c in "աբ":
    print(f"  row {stoi[c]:>2} ({c}): {grad[stoi[c]].tolist()}")

used = (grad.abs().sum(dim=1) > 0).nonzero().flatten()
print(f"\nrows with a nonzero gradient: {[itos[int(i)] for i in used]} - "
      f"the other {V_in - len(used)} of {V_in} rows get exactly 0 and do not move this step")

stand-in loss = sum of the 24 looked-up numbers = -7.844

gradient of the rows that were looked up:
  row 25 (պ): [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
  row 23 (ո): [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
  row 17 (ղ): [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
gradient of two rows that were not:
  row  1 (ա): [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
  row  2 (բ): [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

rows with a nonzero gradient: ['ղ', 'ո', 'պ'] - the other 37 of 40 rows get exactly 0 and do not move this step


### 4.2 The architecture

Follow one context, «պող», through the network (numbers are tensor shapes, batch of 1):

```
  input         պ        ո        ղ              3 letters of context (K = 3)
                |        |        |
  codes         25       23       17             row numbers, not quantities
                |        |        |
  embedding    C[25]    C[23]    C[17]           3 rows of the table C (40 x 8), 8 numbers each
                \        |        /
  concatenate   [ 8 numbers | 8 numbers | 8 numbers ]  ->  24 numbers
                         |
  hidden        ReLU(W1 x + b1)                  24 -> 128 numbers
                         |
  output        W2 h + b2                        128 -> 39 scores ("logits")
                         |
  softmax       39 probabilities, summing to 1:  P(next = ա), P(next = բ), ..., P(next = .)
```

The same thing in PyTorch - `nn.Module` is the standard container: declare the layers in
`__init__`, say how data flows through them in `forward`:

In [17]:
class SurnameMLP(nn.Module):
    def __init__(self, V_in, V_out, K, d, h):
        super().__init__()
        self.emb = nn.Embedding(V_in, d)   # V_in rows: one d-dim vector per readable symbol
        self.fc1 = nn.Linear(K * d, h)     # affine ([41]: weighted sum + bias) ...
        self.fc2 = nn.Linear(h, V_out)     # ... -> ReLU -> affine -> V_out logits

    def forward(self, x):                  # x: (batch, K) integer codes
        e = self.emb(x)                    # (batch, K, d)  look up K vectors
        e = e.view(x.shape[0], -1)         # (batch, K*d)   concatenate them
        h = torch.relu(self.fc1(e))        # (batch, h)     hidden layer
        return self.fc2(h)                 # (batch, V_out) one score per writable symbol

In [18]:
# Walk «պող» through a fresh model step by step, printing the shape at each stage.
model = SurnameMLP(V_in, V_out, K, d=8, h=128)

x = torch.tensor([encode("պող")])                  # a batch of 1 context
e = model.emb(x)
e_flat = e.view(1, -1)
hidden = torch.relu(model.fc1(e_flat))
logits = model.fc2(hidden)
probs = F.softmax(logits, dim=1)

for label, t in [("codes", x), ("embeddings", e), ("concatenated", e_flat),
                 ("hidden (after ReLU)", hidden), ("logits", logits), ("probabilities", probs)]:
    print(f"{label:>20}: shape {tuple(t.shape)}")
print(f"{'':>20}  probabilities sum to {probs.sum().item():.3f}")

               codes: shape (1, 3)
          embeddings: shape (1, 3, 8)
        concatenated: shape (1, 24)
 hidden (after ReLU): shape (1, 128)
              logits: shape (1, 39)
       probabilities: shape (1, 39)
                      probabilities sum to 1.000


**What `.view` does.** The "concatenate" step above is `e.view(1, -1)`. `.view` re-reads the
same numbers in a new shape - nothing is copied, nothing changes, only how the numbers are
grouped. `-1` means "work this size out yourself". A toy first, then our 3 letter vectors:

In [19]:
t = torch.arange(6)
print("t:            ", t.tolist())
print("t.view(2, 3):\n", t.view(2, 3))
print("t.view(3, -1):\n", t.view(3, -1), "   <- -1 became 2, because 6 / 3 = 2")

print("\ne[0]: 3 letter vectors of 8 numbers (պ, ո, ղ):\n", e[0].detach().round(decimals=2))
print("\ne.view(1, -1): the same 24 numbers in one row, պ's 8 then ո's 8 then ղ's 8:\n",
      e_flat.detach().round(decimals=2))

t:             [0, 1, 2, 3, 4, 5]
t.view(2, 3):
 tensor([[0, 1, 2],
        [3, 4, 5]])
t.view(3, -1):
 tensor([[0, 1],
        [2, 3],
        [4, 5]])    <- -1 became 2, because 6 / 3 = 2

e[0]: 3 letter vectors of 8 numbers (պ, ո, ղ):
 tensor([[ 1.5900,  0.0600,  1.1400, -0.2700,  0.2400, -0.8100, -1.9200,  0.1700],
        [ 1.5800, -0.6400, -1.2300, -0.0400,  2.4200, -1.0200,  0.1500,  0.4900],
        [-0.2100,  0.0200,  0.8700, -1.0100, -0.2300,  1.2100,  0.2200,  1.0100]])

e.view(1, -1): the same 24 numbers in one row, պ's 8 then ո's 8 then ղ's 8:
 tensor([[ 1.5900,  0.0600,  1.1400, -0.2700,  0.2400, -0.8100, -1.9200,  0.1700,
          1.5800, -0.6400, -1.2300, -0.0400,  2.4200, -1.0200,  0.1500,  0.4900,
         -0.2100,  0.0200,  0.8700, -1.0100, -0.2300,  1.2100,  0.2200,  1.0100]])


### 4.3 How many parameters?

Count layer by layer: the embedding table, then weights + biases of each linear layer.

$$\underbrace{V_{in}\, d}_{\text{embedding}} \;+\; \underbrace{(K d)\, h + h}_{\text{hidden layer}}
\;+\; \underbrace{h\, V_{out} + V_{out}}_{\text{output layer}}
= 40 \cdot 8 + 24 \cdot 128 + 128 + 128 \cdot 39 + 39 = 8{,}551$$

PyTorch agrees (note that it stores a `Linear` weight as (out, in), so `fc1.weight` prints
as (128, 24)):

In [20]:
for name, p in model.named_parameters():
    print(f"{name:>11}: shape {str(tuple(p.shape)):>9} -> {p.numel():>5} numbers")

d, h = 8, 128
by_formula = V_in * d + (K * d) * h + h + h * V_out + V_out
print(f"\ntotal: {sum(p.numel() for p in model.parameters())}   (formula: {by_formula})")

 emb.weight: shape   (40, 8) ->   320 numbers
 fc1.weight: shape (128, 24) ->  3072 numbers
   fc1.bias: shape    (128,) ->   128 numbers
 fc2.weight: shape (39, 128) ->  4992 numbers
   fc2.bias: shape     (39,) ->    39 numbers

total: 8551   (formula: 8551)


### 4.4 The same network without an embedding layer

What would we build if we skipped `nn.Embedding` and fed the one-hots straight into the MLP?
Each context becomes $K$ one-hots glued together - $3 \times 40 = 120$ numbers, all zeros
except three 1s - followed by the same two linear layers:

```
  պ ո ղ  ->  [0 ... 1 ... 0 | 0 ... 1 ... 0 | 0 ... 1 ... 0]   120 numbers
         ->  ReLU(W1 x + b1)    120 -> 128
         ->  W2 h + b2          128 -> 39 logits
```

In [21]:
class PlainMLP(nn.Module):
    """The same MLP with no embedding layer: K one-hots in, glued together."""
    def __init__(self, V_in, V_out, K, h):
        super().__init__()
        self.V_in = V_in
        self.fc1 = nn.Linear(K * V_in, h)  # 120 inputs (3 one-hots of 40) -> h
        self.fc2 = nn.Linear(h, V_out)

    def forward(self, x):                  # x: (batch, K) integer codes
        onehots = F.one_hot(x, num_classes=self.V_in).float()  # (batch, K, V_in)
        flat = onehots.view(x.shape[0], -1)                     # (batch, K*V_in)  glue them
        return self.fc2(torch.relu(self.fc1(flat)))             # (batch, V_out)   logits


plain = PlainMLP(V_in, V_out, K, h=128)
print(f"plain one-hot MLP: {sum(p.numel() for p in plain.parameters()):>6,} weights")
print(f"embedding MLP:     {sum(p.numel() for p in model.parameters()):>6,} weights")

plain one-hot MLP: 20,519 weights
embedding MLP:      8,551 weights


Nearly all of the difference sits in the first layer. The plain model's `fc1` has one
128-number column for every (symbol, position) pair: 120 columns, so ն in the last slot and
ն in the first slot get unrelated columns. The embedding model describes each letter **once**,
with 8 numbers, and every position reads that same description.

In fact the embedding model *is* a plain MLP - one whose first-layer matrix is forced into a
product. For each position $j$:

$$\underbrace{W^{\text{plain}}_j}_{128 \times 40} \;=\; \underbrace{W^{(1)}_j}_{128 \times 8}\;
\underbrace{C^{\top}}_{8 \times 40}$$

where $C$ is the embedding table and $W^{(1)}_j$ the block of `fc1` that reads position $j$.
Don't take it on faith - build a plain MLP out of the embedding model's weights and compare:

In [22]:
C = model.emb.weight                      # (40, 8)    the symbol table
W1 = model.fc1.weight                     # (128, 24)  3 position blocks of 8 columns
with torch.no_grad():
    # the plain layer's block for position j = W1's block j times the letter table
    blocks = [W1[:, d * j:d * (j + 1)] @ C.T for j in range(K)]     # 3 x (128, 40)
    plain.fc1.weight.copy_(torch.cat(blocks, dim=1))                 # (128, 120)
    plain.fc1.bias.copy_(model.fc1.bias)
    plain.fc2.load_state_dict(model.fc2.state_dict())
    same = torch.allclose(plain(Xtr[:100]), model(Xtr[:100]), atol=1e-5)
print("a plain MLP built from the embedding model's weights gives the same logits:", same)

a plain MLP built from the embedding model's weights gives the same logits: True


So the embedding layer is a **constraint**, not a new ability. What it buys:

1. **One description per letter, shared by all positions** - instead of an unrelated copy for
   each position.
2. **Far fewer weights in the first layer** - $V_{in}d + Kdh$ instead of $KV_{in}h$: 3,392 vs 15,360
   here. With a vocabulary of thousands of words, like Bengio's - let alone the tens of
   thousands of tokens of a modern LLM - the plain first layer becomes enormous; the
   embedding keeps it small. That is why every language model starts with an embedding table.
3. **A table you can look at** - "the vector for ա" is one row (bonus 13 maps it).

Whether the constraint *helps* at 38 letters and 689 names is a different question - an
empirical one. Section 9 trains both and compares.

### 4.5 Where do the starting weights come from?

Before training every weight is random, but not arbitrary: its *spread* is chosen so that
signals neither shrink nor blow up from layer to layer - [45]'s initialization story. The
named recipes, as the standard deviation of the starting weights (fan_in = inputs of the
layer, fan_out = outputs):

| name | std of the weights | made for |
|---|---|---|
| **LeCun** | $\sqrt{1/\text{fan\_in}}$ | tanh-like layers |
| **Glorot / Xavier** | $\sqrt{2/(\text{fan\_in} + \text{fan\_out})}$ | tanh / sigmoid, forward and backward at once |
| **He / Kaiming** | $\sqrt{2/\text{fan\_in}}$ | ReLU layers - the $2/m$ derived in [45] |
| **PyTorch `nn.Linear` default** | $\sqrt{1/(3\,\text{fan\_in})}$, i.e. uniform on $\pm 1/\sqrt{\text{fan\_in}}$ | a general-purpose default |

We keep PyTorch's defaults. `nn.Embedding` starts from a standard normal, $\mathcal{N}(0, 1)$.
`nn.Linear` is a trap for the name-reader: PyTorch implements it as `kaiming_uniform_` with
`a=√5`, and its own source comment explains that this setting is chosen to reproduce the plain
uniform $\pm 1/\sqrt{\text{fan\_in}}$. So despite the "Kaiming" in the code it is **not** the
He recipe of [45]: its spread is $\sqrt{6} \approx 2.4$ times narrower. For a network this
shallow the difference is small; in deep stacks the scale compounds layer after layer, which
is why [45] cares. And `torch.manual_seed(509)` in the setup cell makes these "random" numbers
identical on every run. The four recipes for our two layers, next to what the model drew:

In [23]:
def init_stds(fan_in, fan_out):
    """Standard deviation of the starting weights under each named recipe."""
    return {"PyTorch default": (1 / (3 * fan_in)) ** 0.5, "LeCun": (1 / fan_in) ** 0.5,
            "Glorot": (2 / (fan_in + fan_out)) ** 0.5, "He": (2 / fan_in) ** 0.5}


print(f"{'':<22}" + "".join(f"{k:>16}" for k in init_stds(1, 1)) + f"{'our weights':>13}")
for name, layer in [("fc1", model.fc1), ("fc2", model.fc2)]:
    stds = init_stds(layer.in_features, layer.out_features)
    label = f"{name} ({layer.in_features} in, {layer.out_features} out)"
    print(f"{label:<22}" + "".join(f"{v:>16.3f}" for v in stds.values())
          + f"{layer.weight.std().item():>13.3f}")
print(f"\nemb.weight: mean {model.emb.weight.mean():+.3f}, std {model.emb.weight.std():.3f}   (N(0, 1))")

                       PyTorch default           LeCun          Glorot              He  our weights
fc1 (24 in, 128 out)             0.118           0.204           0.115           0.289        0.119
fc2 (128 in, 39 out)             0.051           0.088           0.109           0.125        0.051

emb.weight: mean +0.005, std 1.029   (N(0, 1))


### 4.6 The untrained net has an opinion

Before any training, feed it the context «պող» and look at its next-character
probabilities. **Predict first:** what *should* a net with random weights believe?

In [24]:
ctx = torch.tensor([encode("պող")])
with torch.no_grad():                       # predicting, not learning -> no gradient tape
    probs = F.softmax(model(ctx), dim=1)[0]

fig = go.Figure(go.Bar(
    x=[itos[i] for i in range(V_out)], y=probs.tolist(), marker_color=BLUE,
    text=[f"{p:.3f}" for p in probs.tolist()], textposition="outside", textangle=-90,
    cliponaxis=False))
fig.add_hline(y=1 / V_out, line=dict(color=GRAY, dash="dash", width=1))
fig.update_layout(title=f"The untrained net's P(next symbol | 'պող'): every bar near the dashed "
                        f"line, uniform 1/{V_out} = {1 / V_out:.3f}",
                  xaxis=dict(title="next symbol", dtick=1),
                  yaxis=dict(title="probability", range=[0, float(probs.max()) * 1.5]),
                  width=920, height=400, template="plotly_white",
                  uniformtext=dict(minsize=8, mode="show"))
fig.show()
print(f"highest {probs.max():.3f}, lowest {probs.min():.3f}, uniform {1 / V_out:.3f}: the output "
      f"weights start small,\nso the logits start small, so the softmax starts close to uniform.")

highest 0.041, lowest 0.017, uniform 0.026: the output weights start small,
so the logits start small, so the softmax starts close to uniform.


## 5. Sampling: asking the net to write

Generation is just the forward pass in a loop: start from an all-`^` context, sample the
next character from the softmax (`torch.multinomial` draws proportionally to probability -
we do not always take the most likely letter, or every name would be the same), slide the
window, and stop when the net emits `.`.

The `temp` argument is a knob we explain in bonus 11; leave it at 1 until then. With
`verbose=True` the function prints every step. We sample from the **untrained** net first -
this is the "before" picture.

In [25]:
@torch.no_grad()                                # generating, not learning: no gradient tape
def sample_names(model, K, n=20, temp=1.0, seed=SEED, verbose=False):
    gen = torch.Generator().manual_seed(seed)   # seeded -> same names on every rerun
    out = []
    for _ in range(n):
        ctx, s = [PAD] * K, ""                  # start: K start tokens = "nothing written yet"
        while True:
            logits = model(torch.tensor([ctx]))[0]
            p = F.softmax(logits / temp if temp > 0 else logits, dim=0)
            if temp == 0:                       # T = 0: always the single most likely char
                i = int(p.argmax())
            else:
                i = int(torch.multinomial(p, 1, generator=gen))   # draw by probability
            if verbose:                         # what the net saw, believed, and drew
                top4 = p.argsort(descending=True)[:4]
                print(f"  '{decode(ctx)}'  top 4: "
                      + "  ".join(f"{itos[int(j)]} {p[j]:.2f}" for j in top4)
                      + f"   -> drew '{itos[i]}' (p = {p[i]:.3f})")
            if i == 0 or len(s) > 25:           # '.' = "done" (25 letters = safety cap)
                break
            s += itos[i]
            ctx = ctx[1:] + [i]                 # slide the window
        out.append(s)
    return out


print("the UNTRAINED net writes:")
print(", ".join(n.capitalize() for n in sample_names(model, K, n=12)))

the UNTRAINED net writes:


Ժիգզբխքհէևլւբծէւսղխշգխզածգ, Խմմցյրծէզչկչեննբքչխգմևքցտս, Սպւձպֆյձիբչէշփօվիէէիեա, Զռօբչթիգէրբւջ, Ծֆքչտշիգծկթջհ, Ւթնզլբջ, Յնչքյսօրչվզքղիտխւհտհղօջձցշ, Ռթնխնսյգսխօվգչշքեյդվիածձ, Գջբժչշիցլսմ, Օաշձզմխշսմվդյռտփկցխցշբզձօհ, Չթսյաեթդնկժչվփմյկբշձխջիտեզ, Չզթսղմֆտյֆհճդք


**One name in slow motion.** Same function, `verbose=True`: at every step, the context the
net saw, its four favourite next characters with their probabilities, and the character it
actually drew. Same seed, so this is the first name from the list above:

In [26]:
name = sample_names(model, K, n=1, verbose=True)[0]
print(f"\n-> «{name.capitalize()}»")

  '^^^'  top 4: բ 0.04  շ 0.03  գ 0.03  հ 0.03   -> drew 'ժ' (p = 0.023)
  '^^ժ'  top 4: բ 0.04  դ 0.04  ց 0.04  վ 0.03   -> drew 'ի' (p = 0.018)
  '^ժի'  top 4: ս 0.04  դ 0.04  յ 0.04  թ 0.03   -> drew 'գ' (p = 0.026)
  'ժիգ'  top 4: խ 0.04  ա 0.04  դ 0.04  ց 0.03   -> drew 'զ' (p = 0.026)
  'իգզ'  top 4: զ 0.04  վ 0.04  խ 0.04  և 0.03   -> drew 'բ' (p = 0.021)
  'գզբ'  top 4: . 0.04  ա 0.04  յ 0.04  ց 0.03   -> drew 'խ' (p = 0.020)
  'զբխ'  top 4: դ 0.05  ջ 0.04  վ 0.04  . 0.03   -> drew 'ք' (p = 0.024)
  'բխք'  top 4: ս 0.05  շ 0.04  ձ 0.04  . 0.03   -> drew 'հ' (p = 0.018)
  'խքհ'  top 4: ա 0.05  օ 0.04  զ 0.03  խ 0.03   -> drew 'է' (p = 0.021)
  'քհէ'  top 4: ա 0.04  բ 0.04  լ 0.04  թ 0.03   -> drew 'և' (p = 0.019)
  'հէև'  top 4: բ 0.04  ա 0.04  զ 0.04  գ 0.03   -> drew 'լ' (p = 0.030)
  'էևլ'  top 4: բ 0.04  ց 0.04  թ 0.04  խ 0.04   -> drew 'ւ' (p = 0.023)
  'ևլւ'  top 4: խ 0.03  կ 0.03  չ 0.03  թ 0.03   -> drew 'բ' (p = 0.027)
  'լւբ'  top 4: . 0.04  գ 0.04  հ 0.03  ց 0.03   ->

Every probability hovers around 1/39 = 0.03, so every draw is close to a coin toss over the
alphabet - and `.` (stop) is just one more unlikely option, which is why an untrained name
tends to run until the 25-letter safety cap. Section 8 replays this for the trained net.

## 6. How good is good? Loss, perplexity, and two baselines

Before training anything, decide how to score a model - and what score "no skill" gets.

### 6.1 The loss: average surprise

For every window the model outputs 39 probabilities. We look only at the probability $p$ it
gave to the character that **actually came next**, and charge $-\log p$ - the model's
*surprise*. The loss is the average surprise over all windows. This is **cross-entropy**,
the same log-loss as [11], now over 39 classes instead of 2.

In [27]:
for p in [1.0, 0.5, 0.2, 1 / V_out, 0.01]:
    print(f"p(correct char) = {p:.3f}  ->  surprise -log p = {np.log(1 / p):.2f}")

p(correct char) = 1.000  ->  surprise -log p = 0.00
p(correct char) = 0.500  ->  surprise -log p = 0.69
p(correct char) = 0.200  ->  surprise -log p = 1.61
p(correct char) = 0.026  ->  surprise -log p = 3.66
p(correct char) = 0.010  ->  surprise -log p = 4.61


Certain and right costs 0; a coin flip costs 0.69; the smaller $p$, the steeper the charge.
Being confidently wrong ($p = 0.01$) costs 4.6 - one such mistake outweighs many good guesses.

### 6.2 Baseline 1: uniform guessing

A model that knows nothing gives each of the $V_{out} = 39$ symbols it can write the same
$1/V_{out}$. Its surprise is $-\log(1/V_{out}) = \log V_{out}$ on every single window, so
that is its loss. Every curve we draw today is read against this line.

In [28]:
uniform_loss = float(np.log(V_out))
print(f"uniform guessing: every symbol gets 1/{V_out}  ->  loss = log({V_out}) = {uniform_loss:.3f}")

uniform guessing: every symbol gets 1/39  ->  loss = log(39) = 3.664


### 6.3 Perplexity: the loss in human units

"Loss 3.66" is hard to feel. **Perplexity** $= e^{\text{loss}}$ undoes the log and answers:
*among how many equally likely letters is the model effectively choosing?* A model that
always narrows the next letter down to $k$ equally likely candidates (one of them right) has
loss $\log k$ - so its perplexity is exactly $k$:

In [29]:
for k in [V_out, 7, 2, 1]:
    print(f"torn between {k:>2} letters: loss log({k}) = {np.log(k):.3f}  ->  "
          f"perplexity e^{np.log(k):.3f} = {np.exp(np.log(k)):.0f}")

torn between 39 letters: loss log(39) = 3.664  ->  perplexity e^3.664 = 39
torn between  7 letters: loss log(7) = 1.946  ->  perplexity e^1.946 = 7
torn between  2 letters: loss log(2) = 0.693  ->  perplexity e^0.693 = 2
torn between  1 letters: loss log(1) = 0.000  ->  perplexity e^0.000 = 1


Uniform guessing: perplexity 39, the whole alphabet. A perfect model: perplexity 1. Every
model today lands somewhere in between.

### 6.4 Baseline 2: the bigram - counting letter pairs

The cheapest model with any memory: **count**, over the training names, how often each
symbol follows each symbol: a table with one row for everything that can come *before*
(`^` or a letter) and one column for everything that can come *after* (a letter or `.`). It
knows exactly one thing - the previous letter - so it is the counting version of a $K = 1$
model.

In [30]:
counts = np.zeros((V_in, V_out))           # counts[a, b] = how often b follows a
for n in tr_names:
    seq = "^" + n + "."                    # ^ marks the start, . the end
    for a, b in zip(seq, seq[1:]):
        counts[stoi[a], stoi[b]] += 1

# Rows that can occur: ^ or a letter. The '.' row stays empty - nothing comes after the end.
before = [stoi["^"]] + [stoi[c] for c in chars]

row = counts[stoi["յ"]]
print(f"'յ' occurs {int(row.sum())} times in the training names; what comes next:")
for j in np.argsort(-row)[:5]:
    print(f"  յ -> {itos[j]}: {int(row[j]):>3} times  ({row[j] / row.sum():.1%})")

'յ' occurs 569 times in the training names; what comes next:
  յ -> ա: 542 times  (95.3%)
  յ -> ո:  10 times  (1.8%)
  յ -> ր:   5 times  (0.9%)
  յ -> ե:   3 times  (0.5%)
  յ -> վ:   2 times  (0.4%)


In [31]:
fig = go.Figure(go.Heatmap(
    z=np.log1p(counts[before]), x=[itos[j] for j in range(V_out)], y=[itos[i] for i in before],
    customdata=counts[before].astype(int).tolist(),   # plain list: plotly's packed format breaks the hover
    colorscale=[[0, "white"], [1, BLUE]], colorbar=dict(title="log(1+count)"),
    hovertemplate="%{y} -> %{x}: %{customdata:.0f} times<extra></extra>"))
fig.update_layout(title="Bigram counts: row = this symbol (^ = start), column = the next one (. = end)",
                  xaxis=dict(title="next symbol", dtick=1),
                  yaxis=dict(title="this symbol", dtick=1, autorange="reversed"),
                  width=720, height=700, template="plotly_white")
fig.show()

Hover over the cells. The «յ» row is nearly one dark cell (յ -> ա, the «-յան» suffix again),
the «^» row shows which letters start names, the «.» column which letters end them - and
most of the table is white: pairs that never occur.

**From counts to probabilities:** divide each row by its total. But the white cells are a
trap: a pair never seen in training gets probability 0, and $-\log 0 = \infty$. A single
unseen pair in the validation names makes the average loss infinite. How many are there?

In [32]:
def bigram_val_loss(P):
    """Average surprise of the probability table P on every letter pair of the val names."""
    surprises = [-np.log(P[stoi[a], stoi[b]])
                 for n in va_names for a, b in zip("^" + n, n + ".")]
    return float(np.mean(surprises)), surprises


print(f"table cells never seen in training: {int((counts[before] == 0).sum())} of {len(before) * V_out}")

P_raw = np.zeros_like(counts)              # the empty '.' row stays zero: it is never looked up
P_raw[before] = counts[before] / counts[before].sum(axis=1, keepdims=True)
with np.errstate(divide="ignore"):         # log(0) = -inf is exactly the point here
    raw_loss, raw_surprises = bigram_val_loss(P_raw)
print(f"validation pairs never seen in training: {int(np.isinf(raw_surprises).sum())} "
      f"of {len(raw_surprises)}")
print(f"raw bigram val loss: {raw_loss}")

table cells never seen in training: 1096 of 1521
validation pairs never seen in training: 48 of 1322
raw bigram val loss: inf


The standard fix is **add-one (Laplace) smoothing**: pretend every pair was seen once more
than it was. Unseen pairs get a small nonzero probability; the price is that frequent pairs
give a little probability away:

In [33]:
P_bigram = np.zeros_like(counts)
P_bigram[before] = (counts[before] + 1) / (counts[before] + 1).sum(axis=1, keepdims=True)   # add-one
a, b = stoi["յ"], stoi["ա"]
print(f"P(ա | յ): raw {P_raw[a, b]:.1%} -> smoothed {P_bigram[a, b]:.1%}")

bigram_loss, _ = bigram_val_loss(P_bigram)
print(f"\nuniform: val loss {uniform_loss:.3f}   perplexity {np.exp(uniform_loss):.1f}")
print(f"bigram:  val loss {bigram_loss:.3f}   perplexity {np.exp(bigram_loss):.1f}")

P(ա | յ): raw 95.3% -> smoothed 89.3%

uniform: val loss 3.664   perplexity 39.0
bigram:  val loss 1.948   perplexity 7.0


In [34]:
# The bigram can write, too: start at '^', draw the next symbol from the current symbol's row.
g_big = np.random.default_rng(SEED)
big_samples = []
for _ in range(10):
    i, s = PAD, ""
    while True:
        i = g_big.choice(V_out, p=P_bigram[i])
        if i == 0 or len(s) > 25:
            break
        s += itos[i]
    big_samples.append(s.capitalize())
print("the bigram model writes:", ", ".join(big_samples))

the bigram model writes: Գւն, Փալբան, Դյապճժրռգրոյակպեցևջաշմադալ, Թեյաջղհանարիվան, Պան, Ան, Եւծկյիրջռին, Բդան, Թևզյագյամասիծշդծիգկունջանի, Վալյարյարուհմյան


One letter of memory buys a lot: from perplexity 39 to about 7. Some outputs are almost
names - but nothing stops a bigram from rambling for 26 characters, because it has no idea
how much name has already happened. That is exactly the job of a longer context window.

So the MLP has to beat **the bigram**, not the clueless 3.66. (A $K = 1$ network learns
essentially this table by gradient descent - [11]'s softmax regression rediscovering the
counts. Section 9 checks that.)

## 7. Training: the loop, for real

This is the standard PyTorch training loop - the one in your ch11 homework, with [42]'s
`loss.backward()` at its core:

```python
for epoch in range(n_epochs):
    for xb, yb in batches:           # mini-batches
        optimizer.zero_grad()        # reset grads
        out = model(xb)              # forward pass
        loss = loss_fn(out, yb)
        loss.backward()              # backprop (autograd runs [42]'s delta recursion)
        optimizer.step()             # update weights
```

- **An epoch** is one pass over all 5,208 training windows, shuffled, in **mini-batches of
  1,024** - so 6 weight updates per epoch.
- **The optimizer is Adam** ([44]): gradient descent with momentum and a per-weight step
  size - the default choice for networks like this one.
- **The loss** is `F.cross_entropy`, which takes the raw **logits**, not softmax outputs: it
  applies log-softmax internally because that is numerically safer. Same math, better
  arithmetic.
- **Early stopping** ([43]) is built in from the start: with 689 names this model overfits
  *fast* - the validation loss bottoms out within the first few dozen epochs and then climbs.
  So `fit()` tracks the best-validation epoch and **keeps those weights**.

In [35]:
def evaluate(model, X, Y):
    """The loss (average cross-entropy) of `model` on windows X with targets Y."""
    with torch.no_grad():
        return float(F.cross_entropy(model(X), Y))


def make_model(config):
    """The network a config describes: the embedding MLP (4.2) or the plain one-hot MLP (4.4)."""
    if config["arch"] == "embedding":
        return SurnameMLP(V_in, V_out, config["K"], config["d"], config["h"])
    if config["arch"] == "plain":
        return PlainMLP(V_in, V_out, config["K"], config["h"])
    raise ValueError(f"unknown arch {config['arch']!r}")


def fit(config, on_epoch=None, seed=SEED):
    """Train the configured network with mini-batch Adam and [43]'s early stopping.
    Returns (best-validation model, loss history, best val loss, best epoch).
    Pure training - pass on_epoch(ep, model, train_loss, val_loss, best_val) to watch it."""
    torch.manual_seed(seed)                          # controls the initial weights
    K = config["K"]
    Xtr, Ytr = build_dataset(tr_names, K)
    Xva, Yva = build_dataset(va_names, K)
    model = make_model(config)
    opt = torch.optim.Adam(model.parameters(), lr=config["lr"])
    g = torch.Generator().manual_seed(seed)          # controls the batch shuffling
    hist = {"train": [], "val": []}
    best_val, best_ep, best_state = float("inf"), 0, None

    for ep in range(1, config["epochs"] + 1):
        # one epoch: a shuffled pass over all training windows, in mini-batches
        idx = torch.randperm(len(Xtr), generator=g)
        for k in range(0, len(idx), config["batch"]):
            b = idx[k:k + config["batch"]]
            loss = F.cross_entropy(model(Xtr[b]), Ytr[b])
            opt.zero_grad()
            loss.backward()
            opt.step()

        tl, vl = evaluate(model, Xtr, Ytr), evaluate(model, Xva, Yva)
        hist["train"].append(tl)
        hist["val"].append(vl)
        if vl < best_val:                            # early stopping, [43]:
            best_val, best_ep = vl, ep               # remember the best epoch's weights
            best_state = copy.deepcopy(model.state_dict())
        if on_epoch:
            on_epoch(ep, model, tl, vl, best_val)

    model.load_state_dict(best_state)                # restore the val-minimum model
    return model, hist, best_val, best_ep

### Weights & Biases in one paragraph

An experiment tracker records, for every **run**: the exact **config** (every knob), the
**metrics** over time (our losses), and any artifacts you log (we will log the *names the
net writes* every few epochs). You then compare runs in a dashboard instead of in your
memory. The discipline it enforces - *no experiment without its settings recorded* - is the
actual lesson; by hand this dies at run 3.

Setup: create a free account at [wandb.ai](https://wandb.ai), copy your API key from
[wandb.ai/authorize](https://wandb.ai/authorize) into the repo-root `.env` as
`WANDB_API_KEY=...` (the setup cell above already loads it) - or run `wandb.login()` once.
**No account / no internet right now?** Uncomment the `WANDB_MODE = "offline"` line at
the top of the next cell: everything still runs and logs to a local folder, nothing is
uploaded - nobody is blocked.

One habit worth keeping forever: before a batch of runs, **check which account will
receive them** - the next cell prints it before anything trains (or confirms that you
are offline). Shared machines accumulate stale logins; runs that land in the wrong
account are annoying to chase.

In [36]:
# ~4 s (asks the wandb server who you are)
# No wandb account, or no internet? Uncomment the next line: runs are then written to a
# local ./wandb folder instead of uploaded, and nothing else in the notebook changes.
# os.environ["WANDB_MODE"] = "offline"

if os.environ.get("WANDB_MODE") == "offline":
    print("WANDB_MODE=offline: runs are logged to ./wandb, nothing is uploaded")
else:
    # Which account gets these runs? A stored login proves *a* login exists, not whose.
    # (An error here means no login at all: run wandb.login(), or go offline above.)
    print("runs will go to wandb entity:", wandb.Api().default_entity)

wandb: [wandb.Api()] Loaded credentials for https://api.wandb.ai from WANDB_API_KEY.


runs will go to wandb entity: tarkhanyan


In [37]:
def cfg_label(c):
    """A readable run name, e.g. 'K3-d8-h128-lr0.003' (or 'plain-K3-h128-lr0.003')."""
    if c["arch"] == "plain":
        return f"plain-K{c['K']}-h{c['h']}-lr{c['lr']}"
    return f"K{c['K']}-d{c['d']}-h{c['h']}-lr{c['lr']}"


def train_logged(config, prefix="main", pause=0.0):
    """fit() + one wandb run that logs the losses every epoch and the names the net writes.
    pause: seconds to wait after each epoch - only so the dashboard can be watched live."""
    K = config["K"]
    # silent=True: none of wandb's own upload chatter in the notebook - we print the link ourselves
    run = wandb.init(project="armenian-name-inventor", name=f"{prefix}-{cfg_label(config)}",
                     config=config, settings=wandb.Settings(silent=True))
    print("wandb run:", run.url or f"offline, saved under {run.dir}")

    def on_epoch(ep, model, tl, vl, best):
        row = {"train_loss": tl, "val_loss": vl, "best_val_loss": best,
               "baseline_uniform": uniform_loss, "baseline_bigram": bigram_loss}
        # Log what the net writes right now. Log-spaced early epochs matter most:
        # the dramatic learning (gibberish -> '-yan') happens in the first ~16 epochs,
        # because the suffix is the single most frequent pattern in the data.
        if ep in (1, 2, 4, 8, 16) or ep % config["sample_every"] == 0:
            tbl = wandb.Table(columns=["name"])
            for s in sample_names(model, K, n=10, seed=SEED + ep):
                tbl.add_data(s.capitalize())
            row["samples"] = tbl
        wandb.log(row, step=ep)
        if ep == 1:                                  # epoch 1 includes wandb's slow first log,
            clock["after_epoch_1"] = time.time()     # so the pace is measured from here on
        elif ep % 10 == 0:                           # progress in the notebook as well
            per_epoch = (time.time() - clock["after_epoch_1"]) / (ep - 1)
            print(f"  epoch {ep:>3}/{config['epochs']}: train {tl:.3f}  val {vl:.3f}  "
                  f"(best {best:.3f})   {time.time() - clock['start']:.0f} s elapsed, "
                  f"~{per_epoch * (config['epochs'] - ep):.0f} s left")
        time.sleep(pause)

    clock = {"start": time.time()}                   # the clock starts with training
    model, hist, best_val, best_ep = fit(config, on_epoch)
    run.finish()
    print(f"best val {best_val:.3f} at epoch {best_ep}/{config['epochs']} "
          f"(uniform {uniform_loss:.3f}, bigram {bigram_loss:.3f}); kept those weights")
    return model, hist, best_ep

### Live: watch it train

This is the cell to run in front of the class. Press run, then **open the link that appears
first**: the run page fills in while the cell is still going - the loss curves grow epoch by
epoch, and the `samples` table gains a new set of names at epochs 1, 2, 4, 8, 16, 25, 50, ...
The model itself trains in about 10 s, far too fast to watch, so `pause=0.5` waits half a
second after each epoch. Set `pause=0` to train at full speed, or raise it for a slower show.

In [38]:
# ~2 min: 150 epochs x 0.5 s pause (so the dashboard can be watched), plus wandb start-up and upload
config = dict(arch="embedding", K=3, d=8, h=128, lr=3e-3, epochs=150, batch=1024, sample_every=25)
model, hist, best_ep = train_logged(config, pause=0.5)

wandb run: https://wandb.ai/tarkhanyan/armenian-name-inventor/runs/9rpvth2w


  epoch  10/150: train 1.824  val 1.912  (best 1.912)   20 s elapsed, ~93 s left


  epoch  20/150: train 1.615  val 1.739  (best 1.739)   26 s elapsed, ~81 s left


  epoch  30/150: train 1.520  val 1.696  (best 1.696)   32 s elapsed, ~74 s left


  epoch  40/150: train 1.453  val 1.679  (best 1.679)   37 s elapsed, ~67 s left


  epoch  50/150: train 1.405  val 1.676  (best 1.674)   44 s elapsed, ~61 s left


  epoch  60/150: train 1.367  val 1.698  (best 1.671)   49 s elapsed, ~54 s left


  epoch  70/150: train 1.327  val 1.706  (best 1.671)   55 s elapsed, ~48 s left


  epoch  80/150: train 1.293  val 1.719  (best 1.671)   61 s elapsed, ~42 s left


  epoch  90/150: train 1.269  val 1.737  (best 1.671)   66 s elapsed, ~35 s left


  epoch 100/150: train 1.244  val 1.752  (best 1.671)   73 s elapsed, ~30 s left


  epoch 110/150: train 1.219  val 1.778  (best 1.671)   79 s elapsed, ~24 s left


  epoch 120/150: train 1.198  val 1.813  (best 1.671)   85 s elapsed, ~18 s left


  epoch 130/150: train 1.179  val 1.839  (best 1.671)   91 s elapsed, ~12 s left


  epoch 140/150: train 1.163  val 1.866  (best 1.671)   97 s elapsed, ~6 s left


  epoch 150/150: train 1.137  val 1.891  (best 1.671)   103 s elapsed, ~0 s left


best val 1.671 at epoch 51/150 (uniform 3.664, bigram 1.948); kept those weights


The same story the dashboard tells, drawn right here:

In [39]:
epochs = list(range(1, config["epochs"] + 1))
fig = go.Figure()
fig.add_scatter(x=epochs, y=hist["train"], name="train loss", line=dict(color=BLUE, width=2))
fig.add_scatter(x=epochs, y=hist["val"], name="validation loss", line=dict(color=RED, width=2))
fig.add_hline(y=uniform_loss, line=dict(color=GRAY, dash="dash", width=1),
              annotation_text="uniform guessing", annotation_position="bottom right")
fig.add_hline(y=bigram_loss, line=dict(color=GRAY, dash="dot", width=1),
              annotation_text="bigram", annotation_position="bottom")
fig.add_vline(x=best_ep, line=dict(color=GRAY, dash="dash", width=1),
              annotation_text=f"best val: epoch {best_ep} (kept)", annotation_position="top right")
fig.update_layout(title="Train keeps falling, validation turns around: early stopping keeps the valley",
                  xaxis_title="epoch", yaxis_title="loss (average cross-entropy)",
                  width=820, height=440, template="plotly_white",
                  legend=dict(x=0.99, y=0.72, xanchor="right"))
fig.show()

## 8. The payoff

Same `sample_names` function as before - but the weights have changed. We ask for 1,000
inventions, look at the first 30, and count over all of them. **Predict first:** what share
will end in «-յան»? (96% of the training names do.)

Then check two more things. How many "inventions" are **real training surnames** repeated
back (memorized - keep that number in mind, it becomes a whole section later)? And the best
kind of hit: how many are **real surnames the net never saw** - names that sit in the
held-out validation set? A model that produces one of those has learned what an Armenian
surname looks like, not just a list of them.

In [40]:
# ~5 s: 1,000 names
invented = sample_names(model, K, n=1000)
print("the TRAINED net writes (the first 30 of 1000):\n")
for i in range(0, 30, 5):
    print("  " + ", ".join(n.capitalize() for n in invented[i:i + 5]))

copied = [n for n in invented if n in tr_set]
rediscovered = [n for n in invented if n in va_set]
print("\nof all 1000 inventions:")
print(f"  end in -յան:                                 {sum(n.endswith('յան') for n in invented)}")
print(f"  copied from the training names:              {len(copied)}")
print(f"  real surnames it never saw (validation set): {len(rediscovered)}  "
      f"{[n.capitalize() for n in rediscovered]}")
print(f"  distinct names:                              {len(set(invented))}")

the TRAINED net writes (the first 30 of 1000):

  Ժիրզախյան, Լիզյան, Անագրյան, Ղամիրյան, Զոյան
  Նիքյան, Բալյան, Աբալյան, Բերշեղյան, Իեանդայբայիլարբաջյան
  Գոշիգոյան, Աբինյան, Իյան, Չելյան, Քղիմյան
  Հարջանիկյան, Նյան, Մավգչշանյան, Ալելգանջինիպիսպարամյան, Դոմյան
  Սոյան, Թազրյան, Թոյան, Դարյան, Մարբանյան
  Թաբչաթսամիկյան, Դալիկյան, Քնդալակյան, Մարյան, Աթրյան

of all 1000 inventions:
  end in -յան:                                 945
  copied from the training names:              48
  real surnames it never saw (validation set): 7  ['Կոսյան', 'Թարյան', 'Կոսյան', 'Բեկյան', 'Աղազարյան', 'Կոսյան', 'Բեկյան']
  distinct names:                              892


945 of the 1,000 end in «-յան» - 94.5%, against 96% in the data: the suffix is learned at
almost exactly the rate the data teaches it. 48 (about 5%) are copied training names; hold on
to that number. And 7 inventions - 4 different names - are **real surnames the net never
saw**, sitting in the held-out validation set: «Կոսյան» (three times), «Բեկյան» (twice),
«Թարյան» and «Աղազարյան». Nobody showed the model those names; it produced them from what
Armenian surnames look like in general. That is generalization in its purest form - and rare
(7 in 1,000), which is why we count over so many names.

The same slow motion as in section 5, now for the trained net - the first invention above,
one step at a time:

In [41]:
name = sample_names(model, K, n=1, verbose=True)[0]
print(f"\n-> «{name.capitalize()}»")

  '^^^'  top 4: ա 0.21  բ 0.10  մ 0.09  թ 0.08   -> drew 'ժ' (p = 0.004)
  '^^ժ'  top 4: ա 0.46  ո 0.27  ե 0.13  ի 0.09   -> drew 'ի' (p = 0.089)
  '^ժի'  top 4: լ 0.28  ր 0.13  ս 0.11  կ 0.08   -> drew 'ր' (p = 0.128)
  'ժիր'  top 4: յ 0.66  ո 0.06  ա 0.05  բ 0.03   -> drew 'զ' (p = 0.023)
  'իրզ'  top 4: ա 0.40  յ 0.31  ո 0.20  ե 0.02   -> drew 'ա' (p = 0.398)
  'րզա'  top 4: ն 0.51  ր 0.18  զ 0.10  յ 0.05   -> drew 'խ' (p = 0.007)
  'զախ'  top 4: յ 0.40  ա 0.34  ի 0.06  բ 0.04   -> drew 'յ' (p = 0.405)
  'ախյ'  top 4: ա 0.99  ո 0.00  ե 0.00  յ 0.00   -> drew 'ա' (p = 0.994)
  'խյա'  top 4: ն 0.99  լ 0.00  ր 0.00  դ 0.00   -> drew 'ն' (p = 0.989)
  'յան'  top 4: . 0.98  ց 0.01  յ 0.00  գ 0.00   -> drew '.' (p = 0.983)

-> «Ժիրզախյան»


Compare with the untrained trace in section 5: the probabilities are now peaked. The first
letters are genuine choices (ա is the favourite start at only 21%), and three times the
sampler drew a long shot - ժ at 0.4%, զ at 2%, խ at 0.7% - which is exactly what sampling is
for, and why this name is unusual. But once a `յ` follows the stem, the rest is nearly
certain: ա 0.994, ն 0.989, then `.` 0.983. The model knows «-յան», and it knows that after
«-յան» the name is over.

Now the ladder. Read it honestly: the humble bigram already covered most of the distance from
clueless to name-shaped; the MLP's edge over it is real but modest, earned on the patterns
longer than one letter. At 689 names, much of "sounding Armenian" *is* letter-pair
statistics.

In [42]:
mlp_loss = evaluate(model, Xva, Yva)
ladder = {"uniform guessing": uniform_loss, "bigram (counts)": bigram_loss, "MLP (ours)": mlp_loss}
for name, loss in ladder.items():
    print(f"{name:>17}: val loss {loss:.3f}   perplexity {np.exp(loss):.1f}")

fig = go.Figure(go.Bar(
    x=[np.exp(v) for v in ladder.values()], y=list(ladder), orientation="h",
    marker_color=[GRAY, GRAY, RED],
    text=[f"perplexity {np.exp(v):.1f}   (loss {v:.2f})" for v in ladder.values()],
    textposition="outside", cliponaxis=False))
fig.update_layout(title="The ladder: among how many letters is each model effectively choosing?",
                  xaxis=dict(title="perplexity (lower is better)", range=[0, 55]),
                  yaxis=dict(autorange="reversed"),
                  width=780, height=300, template="plotly_white", margin=dict(r=60))
fig.show()

 uniform guessing: val loss 3.664   perplexity 39.0
  bigram (counts): val loss 1.948   perplexity 7.0
       MLP (ours): val loss 1.671   perplexity 5.3


### Read the dashboard

Open the run's wandb page (the link printed by the training cell). Three things to find:

1. **Charts -> `train_loss` / `val_loss`** against the two baselines. This is [43]'s
   read-your-curves frame happening live: train loss falls and keeps falling, while the
   validation curve bottoms out within the first few dozen epochs and then **turns upward**
   - the model is memorizing 551 surnames, not learning "Armenian". That early valley is
   exactly why `fit()` kept the best-epoch weights.
2. **The `samples` table**, with its step slider - the first checkpoints are log-spaced
   (epochs 1, 2, 4, 8, 16) because that is where the learning is dramatic. Notice *what*
   the net learns first: the «-յան» ending shows up almost immediately, long before the
   stems sound Armenian. No mystery - 96% of training names carry that suffix, so it is
   the single most loss-reducing pattern available, and gradient descent buys the cheapest
   improvement first. Frequent patterns are learned early; rare structure comes late (or
   never, at this data size).
3. **Overview -> Config** - the exact knobs of this run, recorded without you doing
   anything. That is the point of a tracker.

One more number worth savoring: this model has **more parameters (8,551) than training
windows (5,208)**. Classical statistics says that must end in tears, yet with early stopping
it generalizes fine. Modern deep learning lives almost entirely in this regime; you have just
met it at surname scale.

## 9. Which knob matters? A small comparison

Three questions. What does each extra letter of context actually buy - the measurement that
section 3's argument promised? Does a bigger network help at 689 names? And does the
embedding layer help, against the plain one-hot MLP of section 4.4? This is
**hyperparameter search** - lecture [08]'s machinery on a new model. We vary **one knob at a
time** around the base config:

- context $K \in \{1, 2, 3, 4, 5, 6, 8\}$,
- hidden size $h \in \{32, 128, 512\}$,
- architecture: the embedding MLP or the plain one-hot MLP,

which costs 10 trainings (the base config sits in all three lists), and log all validation
curves into a **single wandb run** (`hp-comparison`): one chart per knob, curves labeled by
value.

Two honest details:

- **The score is the best val loss reached during the run** - early stopping applied to the
  comparison itself. Final val loss would punish big models for overshooting their early
  minimum, not for being bad. The flip side: a run whose best epoch is at the very end of the
  budget was still improving, so its score is capped by the budget. The leaderboard flags it.
- **How big a difference is real?** We retrain the base config with two more seeds: the
  spread between identical configs is the noise floor. A gap smaller than that is not a
  finding.

At real scale you would not do this by hand: **wandb sweeps** automate random search ([08]:
beats grid for the same budget) across many machines. Same idea, more machinery.

In [43]:
# ~1-2 min: 10 trainings of 100 epochs (44-103 s measured)
base = dict(arch="embedding", K=3, d=8, h=128, lr=3e-3, epochs=100, batch=1024)
knobs = {"K": [1, 2, 3, 4, 5, 6, 8], "h": [32, 128, 512], "arch": ["embedding", "plain"]}

run = wandb.init(project="armenian-name-inventor", name="hp-comparison",
                 config={"base": base, "knobs": knobs}, settings=wandb.Settings(silent=True))
print("wandb run:", run.url or f"offline, saved under {run.dir}")

results = {}                    # cfg_label -> (config, history, best_val, best_ep, trained model)
t0 = time.time()
for knob, values in knobs.items():
    curves, labels = [], []
    for v in values:
        cfg = dict(base, **{knob: v})      # the base config with one knob changed
        key = cfg_label(cfg)
        already_trained = key in results   # the base config appears in every list
        if not already_trained:
            model_k, hist_k, best_val_k, best_ep_k = fit(cfg)
            results[key] = (cfg, hist_k, best_val_k, best_ep_k, model_k)
        _, hist_k, best_val_k, best_ep_k, _ = results[key]
        curves.append(hist_k["val"])
        labels.append(f"{knob}={v}")
        note = "(base config, trained above)" if already_trained else f"({time.time() - t0:.0f} s elapsed)"
        print(f"  {knob}={v}: best val {best_val_k:.3f} at epoch {best_ep_k}   {note}")
    run.log({f"varying {knob}": wandb.plot.line_series(
        xs=list(range(1, base["epochs"] + 1)), ys=curves, keys=labels,
        title=f"val loss vs epoch - varying {knob}", xname="epoch")})

wandb run: https://wandb.ai/tarkhanyan/armenian-name-inventor/runs/fgebfj17


  K=1: best val 1.901 at epoch 40   (9 s elapsed)


  K=2: best val 1.713 at epoch 51   (19 s elapsed)


  K=3: best val 1.671 at epoch 51   (30 s elapsed)


  K=4: best val 1.663 at epoch 51   (45 s elapsed)


  K=5: best val 1.658 at epoch 41   (59 s elapsed)


  K=6: best val 1.675 at epoch 32   (71 s elapsed)


  K=8: best val 1.682 at epoch 32   (78 s elapsed)


  h=32: best val 1.655 at epoch 75   (84 s elapsed)
  h=128: best val 1.671 at epoch 51   (base config, trained above)


  h=512: best val 1.713 at epoch 22   (95 s elapsed)


  arch=embedding: best val 1.671 at epoch 51   (base config, trained above)


  arch=plain: best val 1.646 at epoch 33   (99 s elapsed)


In [44]:
# ~10 s: 2 more trainings
# The noise floor: the SAME base config, two more seeds (other initial weights, other batch order)
seed_vals = [results[cfg_label(base)][2]]
for s in [SEED + 1, SEED + 2]:
    seed_vals.append(fit(base, seed=s)[2])
noise = max(seed_vals) - min(seed_vals)
print("base config, 3 seeds:", "  ".join(f"{v:.3f}" for v in seed_vals), f"  -> spread {noise:.3f}")

base config, 3 seeds: 1.671  1.669  1.664   -> spread 0.007


In [45]:
# ~5 s (wandb uploads the finished comparison run)
def n_params(cfg):
    return sum(p.numel() for p in make_model(cfg).parameters())


leaderboard = sorted(results.values(), key=lambda r: r[2])     # best val first
tbl = wandb.Table(columns=["config", "params", "best_val", "best_epoch"])
print(f"{'config':>22} {'params':>7} {'best val':>9} {'epoch':>6}")
for cfg, _, best_val_k, best_ep_k, _ in leaderboard:
    flag = "  <- best epoch at the end of the budget: may still be improving" \
        if best_ep_k > 0.9 * cfg["epochs"] else ""
    print(f"{cfg_label(cfg):>22} {n_params(cfg):>7,} {best_val_k:>9.3f} {best_ep_k:>6}{flag}")
    tbl.add_data(cfg_label(cfg), n_params(cfg), round(best_val_k, 3), best_ep_k)
print(f"\nnoise floor (seed spread): {noise:.3f}     bigram baseline: {bigram_loss:.3f}")

run.log({"leaderboard": tbl})
run.finish()

                config  params  best val  epoch
 plain-K3-h128-lr0.003  20,519     1.646     33
     K3-d8-h32-lr0.003   2,407     1.655     75
    K5-d8-h128-lr0.003  10,599     1.658     41
    K4-d8-h128-lr0.003   9,575     1.663     51
    K3-d8-h128-lr0.003   8,551     1.671     51
    K6-d8-h128-lr0.003  11,623     1.675     32
    K8-d8-h128-lr0.003  13,671     1.682     32
    K3-d8-h512-lr0.003  33,127     1.713     22
    K2-d8-h128-lr0.003   7,527     1.713     51
    K1-d8-h128-lr0.003   6,503     1.901     40

noise floor (seed spread): 0.007     bigram baseline: 1.948


The context knob deserves its own picture - $K$ is ordered, so a line through the seven
runs shows what each extra letter buys:

In [46]:
k_runs = [results[cfg_label(dict(base, K=k))] for k in knobs["K"]]
fig = go.Figure(go.Scatter(
    x=knobs["K"], y=[r[2] for r in k_runs], mode="lines+markers+text",
    text=[f"{r[2]:.3f}" for r in k_runs], cliponaxis=False,
    textposition=["bottom left" if k == 2 else "top center" for k in knobs["K"]],
    line=dict(color=BLUE, width=2), marker=dict(size=9)))
fig.add_hline(y=bigram_loss, line=dict(color=GRAY, dash="dot", width=1),
              annotation_text="bigram", annotation_position="bottom right")
fig.update_layout(title=f"Best validation loss vs context size K (seed noise {noise:.3f})",
                  xaxis=dict(title="K = letters of context", dtick=1),
                  yaxis=dict(title="best val loss (lower is better)", range=[1.6, 2.0]),
                  width=780, height=420, template="plotly_white")
fig.show()

And, because val loss is only a proxy for what we actually want, what each model *writes* -
the same seed for every model, so the rows are directly comparable:

In [47]:
for k in knobs["K"]:
    model_k = results[cfg_label(dict(base, K=k))][4]
    print(f"K={k}:     " + ", ".join(n.capitalize() for n in sample_names(model_k, k, n=6)))
plain_model = results[cfg_label(dict(base, arch="plain"))][4]
print("plain K=3: " + ", ".join(n.capitalize() for n in sample_names(plain_model, 3, n=6)))

K=1:     Ժիրզախյայալիբծերսան, Գազան, Ղամարյան, Զյան, Սնցան, Գյան
K=2:     Ժիրզարյան, Լաջյան, Ան, Գազան, Ղամիրյան, Զոյան
K=3:     Ժիրզախյան, Լիզյան, Անագրյան, Ղամիրյան, Զոյան, Նիքյան
K=4:     Ժիրզարյան, Ազարյան, Ասիրյան, Ղամիրյան, Զոյան, Նիքյան
K=5:     Ժիրզարյան, Ազարյան, Ասերյան, Ղամիրյան, Զովուննցան, Գյուլյան
K=6:     Ժիրզարյան, Ազարյան, Ասերյան, Ղամիրյան, Զոյան, Նբալիբյան
K=8:     Ժիրզարյան, Ազարյան, Աշերյան, Ղխմիրյան, Զոյան, Նբալիյան
plain K=3: Ժիրզախվայան, Բարգսյան, Բալանյան, Բերթումյան, Նեքյան, Բալյան


Read it against the noise floor - three seeds of the same base config spread by 0.007:

- **K = 1 lands at 1.901, next to the bigram (1.948)** - the check promised in section 6: one
  letter of context, counted or learned, gets you to the same place.
- **The big steps are K = 1 -> 2 -> 3** (1.901 -> 1.713 -> 1.671). **After K = 3 the curve
  is flat within the noise**: K = 4 and 5 come out a little lower (1.663, 1.658 - one and two
  noise floors), K = 6 and 8 a little higher (1.675, 1.682). Longer windows buy nothing you
  could bet on. The names agree: K = 1 and 2 write fragments like «Գազան», «Ան», «Զյան»; from
  K = 3 on every row sounds like surnames, with no row clearly better than the others.
- **h = 512 is worse (1.713)**: 4x the weights, it memorizes faster (best epoch 22) and lands
  higher. **h = 32 is better (1.655)** with a quarter of the base model's weights - at 689
  names, *less* capacity helps a little.
- **The plain one-hot MLP tops the board (1.646)**, 0.025 below the embedding model - about
  3.5 noise floors. At 38 letters the embedding's constraint costs a little: the plain model
  may give ն at the end of the window a different role than ն at the start, and with only 120
  inputs its extra weights are affordable. The constraint pays off when the vocabulary is
  large (section 4.4) - not here.

The honest verdict: none of these knobs moves much. Our base config is 0.025 behind the best
row, next to the 0.230 it gained over K = 1 and the 1.993 it gained over uniform guessing.
At 689 names **the data is the bottleneck**, not the architecture - which is, in one
sentence, why the big-model era is above all a big-data era.

**So which model do we keep?** On the numbers alone, the plain MLP (or h = 32). We keep the
base config anyway, and say so openly: the winners are ahead by about 1.5% of the loss, while
the embedding model with K = 3 is the one this whole notebook explains, the architecture
every language model uses, and the one whose letter table bonus 13 inspects. The model
trained in section 7 already *is* that config, so it becomes our final model. And val loss
is only a **proxy** for what we want (charming names) - the rows of names above are the
other half of the evidence.

## 10. Save the model, load it back

Training took seconds here; for a real model it takes weeks, and nobody retrains a model to
use it. So what has to go into the file? Three things - and forgetting either of the last two
is the classic mistake:

1. **The weights** - `model.state_dict()`, a dict of named tensors.
2. **The config** ($K$, $d$, $h$) - you must rebuild an empty model of the same shape before
   the weights can be poured into it.
3. **The vocabulary** (`stoi`) - the weights of "row 25" mean nothing unless row 25 still
   means պ. Rebuild the alphabet from a different name list and every row shifts.

We save the `state_dict`, not the model object: `torch.save(model)` pickles your class by its
import path, so the file breaks the day you rename or move the class. A `state_dict` is just
tensors.

In [48]:
ckpt_path = Path("data/name_inventor.pt")
checkpoint = {
    "state_dict": {k: v.cpu() for k, v in model.state_dict().items()},   # CPU tensors load anywhere
    "config": config,
    "stoi": stoi,
}
torch.save(checkpoint, ckpt_path)

print(f"saved {ckpt_path} ({ckpt_path.stat().st_size / 1024:.0f} KB)")
print("inside:", list(checkpoint))
print("weights:", {k: tuple(v.shape) for k, v in checkpoint["state_dict"].items()})

saved data\name_inventor.pt (37 KB)
inside: ['state_dict', 'config', 'stoi']
weights: {'emb.weight': (40, 8), 'fc1.weight': (128, 24), 'fc1.bias': (128,), 'fc2.weight': (39, 128), 'fc2.bias': (39,)}


Now pretend it is next week and this is a fresh Python session. Loading is the save in
reverse: read the file, build an empty model from the config, pour the weights in. Two safety
details:

- `weights_only=True` (the default since PyTorch 2.6) refuses to run code hidden in the file.
  A `.pt` file is a pickle, and a pickle from the internet could otherwise execute anything
  on load.
- `load_state_dict` is **strict**: a missing or misshaped tensor raises an error instead of
  quietly loading a broken model.

In [49]:
ckpt = torch.load(ckpt_path, weights_only=True)
assert ckpt["stoi"] == stoi, "alphabet changed since saving - the rows would mean other letters"

cfg = ckpt["config"]
loaded = make_model(cfg)                                            # empty model, same shape
loaded.load_state_dict(ckpt["state_dict"])                          # strict: shapes must match
loaded.eval()        # inference mode - a no-op here, vital once a model has dropout/BatchNorm ([46])

print(f"val loss:  original {evaluate(model, Xva, Yva):.4f}   loaded {evaluate(loaded, Xva, Yva):.4f}")
same = sample_names(loaded, cfg["K"], n=10) == sample_names(model, K, n=10)
print("same 10 names from both:", same)

# From here on, the bonuses use the model we just loaded.
final_model, final_K = loaded, cfg["K"]

val loss:  original 1.6710   loaded 1.6710
same 10 names from both: True


## 11. Bonus: the temperature knob

Every chat app has a *temperature* setting. Here is all it is: divide the logits $z$ by a
number $T$ before the softmax.

$$p_i = \frac{e^{z_i / T}}{\sum_j e^{z_j / T}}$$

- **$T = 1$** - the model's own distribution; everything we sampled so far.
- **$T < 1$** - the gaps between logits grow, so the favourite gets more and the long tail
  less: safer, more typical output.
- **$T > 1$** - the gaps shrink and the distribution flattens: wilder output, more nonsense.
  As $T \to \infty$ it becomes uniform - the model's knowledge is erased.
- **$T = 0$** - the limit of sharpening: all probability on the single top choice, no
  randomness left. This is **greedy decoding** (argmax). Dividing by 0 is undefined, so code
  treats $T = 0$ as argmax - exactly what `sample_names` does.

Three logits make it concrete:

In [50]:
z = torch.tensor([2.0, 1.0, 0.0])          # logits for three letters, say ա, բ, գ
print("logits:", z.tolist(), "\n")
print("T = 0    ->", F.one_hot(z.argmax(), 3).float().tolist(), "  (argmax: all on the top letter)")
for T in [0.5, 1.0, 2.0, 100.0]:
    p = F.softmax(z / T, dim=0)
    print(f"T = {T:<5} ->", [round(v, 3) for v in p.tolist()])

logits: [2.0, 1.0, 0.0] 

T = 0    -> [1.0, 0.0, 0.0]   (argmax: all on the top letter)
T = 0.5   -> [0.867, 0.117, 0.016]
T = 1.0   -> [0.665, 0.245, 0.09]
T = 2.0   -> [0.506, 0.307, 0.186]
T = 100.0 -> [0.337, 0.333, 0.33]


Now the real model. Which letter should a surname **start** with? That is one softmax - the
context `^^^` - shown at four temperatures (the 8 most likely first letters):

In [51]:
with torch.no_grad():
    z = final_model(torch.tensor([[PAD] * final_K]))[0]   # logits for the first letter
top = z.argsort(descending=True)[:8]                       # the 8 most likely first letters
temps = [0, 0.5, 1.0, 2.0]

fig = make_subplots(rows=1, cols=4, shared_yaxes=True, horizontal_spacing=0.03,
                    subplot_titles=[f"T = {T}" + (" (greedy)" if T == 0 else "") for T in temps])
for col, T in enumerate(temps, start=1):
    p = F.one_hot(z.argmax(), V_out).float() if T == 0 else F.softmax(z / T, dim=0)
    fig.add_bar(x=[itos[int(i)] for i in top], y=p[top].tolist(), marker_color=BLUE,
                text=[f"{v:.0%}" for v in p[top].tolist()], textposition="outside",
                cliponaxis=False, showlegend=False, row=1, col=col)
    print(f"T = {T}: the top 8 letters hold {p[top].sum():.0%} of the probability")
fig.update_yaxes(range=[0, 1.12], tickformat=".0%")
fig.update_yaxes(title_text="P(first letter)", row=1, col=1)
fig.update_layout(title="The same logits at four temperatures: sharper below 1, flatter above",
                  width=960, height=380, template="plotly_white",
                  uniformtext=dict(minsize=10, mode="show"))   # same label size on every bar
fig.show()

T = 0: the top 8 letters hold 100% of the probability
T = 0.5: the top 8 letters hold 95% of the probability
T = 1.0: the top 8 letters hold 70% of the probability
T = 2.0: the top 8 letters hold 47% of the probability


And the names themselves. We use a fresh seed, so the $T = 1$ row is new names rather than
a rerun of section 8's list, and 200 names per temperature so the counts mean something:

In [52]:
# ~5 s: 600 names
greedy = sample_names(final_model, final_K, n=5, temp=0)
print("T = 0 (greedy):", ", ".join(n.capitalize() for n in greedy))
print(f"   -> the same name every time. In the training names: {greedy[0] in tr_set}, "
      f"in the validation names: {greedy[0] in va_set}\n")

for T in [0.5, 1.0, 1.5]:
    s = sample_names(final_model, final_K, n=200, temp=T, seed=SEED + 1)
    print(f"T = {T}: copied from training {sum(x in tr_set for x in s):>2}/200 | "
          f"distinct {len(set(s)):>3}/200 | mean length {np.mean([len(x) for x in s]):.1f}")
    print("   " + ", ".join(n.capitalize() for n in s[:8]) + "\n")

T = 0 (greedy): Անարյան, Անարյան, Անարյան, Անարյան, Անարյան
   -> the same name every time. In the training names: False, in the validation names: False



T = 0.5: copied from training 29/200 | distinct 160/200 | mean length 7.3
   Պամյան, Արիբյան, Բունյան, Ադանյան, Աբիկյան, Արաթյան, Հարաջյան, Ավանյան



T = 1.0: copied from training  9/200 | distinct 192/200 | mean length 8.4
   Պամյան, Թրվևիրանյան, Մատյան, Ազուրյան, Գալալիկաթյան, Հարծրդեսյան, Օմոյան, Թոյան



T = 1.5: copied from training  2/200 | distinct 199/200 | mean length 12.6
   Պավդալզյան, Հրավյան, Ոդտյան, Ազուրյանգրաբալրաթյայան, Մավրդիշոսյան, Մանյանց, Ժահիյան, Մեյան



Read the chart and the counts together:

- **T = 0** writes «Անարյան» every single time - the one most likely surname, and not a real
  one (it is in neither name list). Greedy decoding is for when you want *the* answer, the
  same every time; it can never surprise you.
- **T = 0.5** plays it safe: short names (7.3 letters on average), only 160 distinct out of
  200, and 29 copied from the training set - about three times as many as at T = 1. Low
  temperature amplifies memorization: the safest name is one the model has already seen.
- **T = 1.0** is the model's own distribution: 192 distinct names, 9 copies.
- **T = 1.5** gambles: almost every name is new (199 distinct, 2 copies), but the mean length
  jumps to 12.6 letters and the stems fall apart («Ազուրյանգրաբալրաթյայան»).

Every generative product you have used ships this exact knob. Now you know it is one
division.

## 12. Bonus: overfitting is memorization

**Predict first:** we train the same model **8x longer** than section 7, with no early
stopping. Better names?

Every 100 epochs we measure two things: the val loss, and the share of 50 freshly generated
names that are **verbatim training names**. One subtlety: even a good model produces *some*
real names by chance (short, high-probability ones) - so the verdict is not "any real name =
guilty" but the **trajectory**: does the copied share climb while the val loss goes the
wrong way?

In [53]:
# ~40-80 s: 1200 epochs, with a progress line every 100
# A raw loop on purpose - NO early stopping here. We *want* to watch it overtrain.
torch.manual_seed(SEED)
K = 3
Xtr, Ytr = build_dataset(tr_names, K)
Xva, Yva = build_dataset(va_names, K)
m_long = SurnameMLP(V_in, V_out, K, 8, 128)
opt = torch.optim.Adam(m_long.parameters(), lr=3e-3)
g = torch.Generator().manual_seed(SEED)

epochs_ax, val_ax, copied_ax = [], [], []
t0 = time.time()
for ep in range(1, 1201):
    idx = torch.randperm(len(Xtr), generator=g)
    for k in range(0, len(idx), 1024):
        b = idx[k:k + 1024]
        loss = F.cross_entropy(m_long(Xtr[b]), Ytr[b])
        opt.zero_grad()
        loss.backward()
        opt.step()
    if ep % 100 == 0:                         # checkpoint: val loss + 50 fresh names
        s = sample_names(m_long, K, n=50, seed=SEED + ep)
        epochs_ax.append(ep)
        val_ax.append(evaluate(m_long, Xva, Yva))
        copied_ax.append(sum(x in tr_set for x in s) / 50)
        elapsed = time.time() - t0
        print(f"epoch {ep:>4}: val loss {val_ax[-1]:.3f}, copied {copied_ax[-1]:>4.0%}   "
              f"({elapsed:.0f} s elapsed, ~{elapsed / ep * (1200 - ep):.0f} s left)")

epoch  100: val loss 1.752, copied   0%   (6 s elapsed, ~67 s left)


epoch  200: val loss 2.019, copied   6%   (11 s elapsed, ~55 s left)


epoch  300: val loss 2.265, copied  12%   (16 s elapsed, ~47 s left)


epoch  400: val loss 2.490, copied  24%   (20 s elapsed, ~41 s left)


epoch  500: val loss 2.662, copied  22%   (25 s elapsed, ~35 s left)


epoch  600: val loss 2.798, copied  24%   (29 s elapsed, ~29 s left)


epoch  700: val loss 2.925, copied  34%   (34 s elapsed, ~24 s left)


epoch  800: val loss 2.965, copied  28%   (41 s elapsed, ~21 s left)


epoch  900: val loss 3.053, copied  50%   (51 s elapsed, ~17 s left)


epoch 1000: val loss 3.116, copied  30%   (61 s elapsed, ~12 s left)


epoch 1100: val loss 3.147, copied  40%   (69 s elapsed, ~6 s left)


epoch 1200: val loss 3.197, copied  32%   (79 s elapsed, ~0 s left)


In [54]:
fig = make_subplots(rows=2, cols=1, shared_xaxes=True, vertical_spacing=0.1,
                    subplot_titles=["validation loss (lower is better)",
                                    "share of generated names copied from the training set"])
fig.add_scatter(x=epochs_ax, y=val_ax, mode="lines+markers", line=dict(color=BLUE, width=2),
                row=1, col=1)
fig.add_hline(y=uniform_loss, line=dict(color=GRAY, dash="dash", width=1), row=1, col=1,
              annotation_text="uniform guessing", annotation_position="bottom right")
fig.add_hline(y=bigram_loss, line=dict(color=GRAY, dash="dot", width=1), row=1, col=1,
              annotation_text="bigram", annotation_position="bottom right")
fig.add_scatter(x=epochs_ax, y=copied_ax, mode="lines+markers", line=dict(color=RED, width=2),
                row=2, col=1)
fig.update_yaxes(range=[1.5, 4.0], row=1, col=1)     # headroom above the uniform line
fig.update_yaxes(tickformat=".0%", range=[0, 1], row=2, col=1)
fig.update_xaxes(title_text="epoch", row=2, col=1)
fig.update_layout(title="Train longer: memorization up, generalization down", showlegend=False,
                  width=820, height=580, template="plotly_white")
fig.show()

print(f"copied share: {copied_ax[0]:.0%} (epoch 100)  ->  {copied_ax[-1]:.0%} (epoch 1200)")
print(f"val loss:     {val_ax[0]:.3f} (epoch 100)  ->  {val_ax[-1]:.3f} (epoch 1200)"
      f"   (bigram {bigram_loss:.3f}, uniform {uniform_loss:.3f})")

copied share: 0% (epoch 100)  ->  32% (epoch 1200)
val loss:     1.752 (epoch 100)  ->  3.197 (epoch 1200)   (bigram 1.948, uniform 3.664)


The trajectory is unambiguous. By epoch 200 the network is already **worse on new names than
the counting bigram** (2.019 vs 1.948), and by epoch 1200 its val loss (3.197) has handed
back most of its advantage over uniform guessing (3.664). Meanwhile the share of copied names
climbs from 0% to 32%, touching 50% on the way. The copied curve is noisy - each checkpoint
samples only 50 names - so read its drift, not its wiggles. (Train even longer and such a
model can end up *worse than uniform guessing* on unseen names while reciting the seen ones perfectly. Not a stupid
model: a model that spent its capacity on the wrong job.)

This is the modern face of [43]'s overfitting lesson: in a generative model, overfitting *is*
memorization - the same failure that makes large language models occasionally recite their
training data (a real privacy issue). The cure is the same as [43]'s: **early stopping** at
the val minimum, which is exactly what `fit()` did.

## 13. Bonus: what did the embeddings learn?

Each character became a learned 8-number vector - a row of the embedding table from
section 4.1. The net was never told anything about phonetics; anything phonetic we find, it
inferred purely from which characters are statistically interchangeable inside 689 surnames.
Two ways to look:

**1. Nearest neighbours in the full 8-D space** - no projection, no distortion: for each
letter, which letters got the most similar vectors (cosine similarity)?

In [55]:
# Every symbol the model reads: ^ and the 38 letters. The '.' row is left out - a name is over
# at '.', so that row is never read and never trained (section 2).
readable = [stoi["^"]] + [stoi[c] for c in chars]
E = final_model.emb.weight.detach().numpy()[readable]    # (39, 8): one row per readable symbol
sym = [itos[i] for i in readable]                        # row number -> symbol
En = E / np.linalg.norm(E, axis=1, keepdims=True)
sim = En @ En.T                                          # cosine similarity, symbol x symbol

print("nearest neighbours in the full 8-D embedding space:")
for ch in "աեէիոօսնր":
    r = sym.index(ch)
    nearest = np.argsort(-sim[r])[1:5]                   # [0] is the letter itself
    print(f"  {ch}:  " + "  ".join(sym[j] for j in nearest))

nearest neighbours in the full 8-D embedding space:
  ա:  ի  ն  զ  և
  ե:  ք  օ  ո  ֆ
  է:  ձ  վ  ճ  հ
  ի:  ց  և  ձ  ա
  ո:  գ  ֆ  ռ  ք
  օ:  ք  ս  ո  ե
  ս:  ծ  օ  մ  կ
  ն:  ա  շ  զ  ժ
  ր:  ճ  մ  լ  դ


**2. A 2-D map with UMAP** ([36] - your own tool from ch10). UMAP builds a neighbour graph in
the 8-D space and lays it out in 2-D so that neighbours stay neighbours. With only 39 points
we shrink its neighbourhood (`n_neighbors=8` instead of the default 15) and use cosine
distance, the same similarity as the table above.

In [56]:
# ~1.5-3 min on the instructor laptop (78-166 s measured), and none of it is the 39 points:
# importing umap also loads TensorFlow (~17 s) and numba-compiles UMAP's code (~30 s),
# and the first fit compiles more.
import warnings
from tqdm import TqdmWarning

with warnings.catch_warnings():            # umap imports tqdm, which warns that the notebook
    warnings.simplefilter("ignore", TqdmWarning)   # progress-bar widget is missing; we never use it
    import umap

xy = umap.UMAP(n_neighbors=8, min_dist=0.3, metric="cosine",
               random_state=SEED, n_jobs=1).fit_transform(E)

VOWELS = set("աեէիոօ")                  # (ը never occurs in the 689 surnames)
kind = ["start token" if c == "^" else "vowel" if c in VOWELS else "consonant" for c in sym]

fig = go.Figure()
for group, color in [("vowel", RED), ("consonant", BLUE), ("start token", ORANGE)]:
    idx = [r for r in range(len(sym)) if kind[r] == group]
    fig.add_scatter(x=xy[idx, 0], y=xy[idx, 1], mode="markers+text", name=group,
                    text=[sym[r] for r in idx], textposition="top center",
                    textfont=dict(size=16), marker=dict(color=color, size=10),
                    hovertemplate="%{text}<extra></extra>")
fig.update_layout(title="Learned character embeddings, UMAP to 2-D",
                  xaxis_title="UMAP 1", yaxis_title="UMAP 2",
                  width=740, height=620, template="plotly_white")
fig.show()

Read both honestly:

- **ա and ն** are mutual close neighbours (ա is ն's nearest, ն is ա's second), and ա and ի
  find each other too. No phonetics explains ա-ն - most likely both vectors are shaped by the
  one overwhelming pattern, «-յան». Similar vectors mean "the model treats them alike", not
  "they sound alike".
- The near-homophones do **not** find each other: ե and է share no neighbours at all, and
  ո and օ only barely (ո is օ's third). There is **no vowel/consonant split** either.

A finding, not a failure. Karpathy's *makemore*, trained on 32,000 English names, does see
the vowels a, e, i, o, u cluster in exactly this kind of plot. Representation quality is
bought with data - the same lesson the near-tie leaderboard taught, now seen from inside the
network. And a caution about the map itself: with only 39 points, UMAP's layout depends on
`n_neighbors` and the seed, and at this data size even the neighbour table depends on the
model's random start. The table is the ground truth for *this* model; the map is the overview.

## 14. Bonus: steer it - names from your initials

Conditioning, the cheapest way: pre-load the context with characters of your choice and let
the net finish. (The grown-up version of this idea is called *prompting*.)

In [57]:
@torch.no_grad()
def finish(model, K, prefix, n=6, temp=1.0, seed=SEED):
    """Like sample_names, but the name starts with `prefix` instead of from scratch."""
    gen = torch.Generator().manual_seed(seed)
    out = []
    for _ in range(n):
        ctx = ([PAD] * K + encode(prefix))[-K:]   # the last K letters of the prefix
        s = prefix.lower()
        while True:
            p = F.softmax(model(torch.tensor([ctx]))[0] / temp, dim=0)
            i = int(torch.multinomial(p, 1, generator=gen))
            if i == 0 or len(s) > 25:
                break
            s += itos[i]
            ctx = ctx[1:] + [i]
        out.append(s.capitalize())
    return out


for prefix in ["տ", "քո", "մաշ"]:          # try your own initials here
    print(f"{prefix}...  ->  " + ", ".join(finish(final_model, final_K, prefix)))

տ...  ->  Տեֆրաբուրյան, Տոչուսայան, Տիլյան, Տոնյան, Տունյան, Տոսակիբյան
քո...  ->  Քոյան, Քոբուրյան, Քոչյան, Քորյան, Քոյան, Քոյամիրյան
մաշ...  ->  Մաշյան, Մաշախյան, Մաշիկյան, Մաշյան, Մաշյան, Մաշյաղիմյան


And a prefix the model cannot read. ը never appears in the 689 surnames (section 2), so it
has no code and no embedding row. `encode` stops with a clear error rather than quietly
dropping the letter and steering with a different prefix than the one you typed:

In [58]:
try:
    finish(final_model, final_K, "ըն")
except ValueError as err:                  # shown here on purpose; normally, let it crash
    print("ValueError:", err)

ValueError: ['ը'] never appear in the 689 surnames, so the model has no code (and no embedding row) for them. Its alphabet: աբգդեզէթժիլխծկհձղճմյնշոչպջռսվտրցւփքօֆև


## 15. What you actually built

A character-level **language model**: embeddings -> MLP -> softmax over the next token,
trained with cross-entropy, scored against honest baselines, saved to a file and loaded back.
That is Bengio et al. (2003) - the direct ancestor of every modern LLM. What separates it
from ChatGPT is scale and one architectural idea: **much** more data, a **much** longer
context, and *attention* in place of our fixed $K$-window (a later chapter). The training
loop, the loss, the sampling, the temperature knob, the memorization risk - those you have
now seen for real, at surname scale.

And keep the wandb habit: from here on, every experiment in this course is a run with its
config recorded.

**Next:** convolutional networks - what happens when the *input* has shape worth
respecting.